---
title: "Native Plants Analysis"
output: html_notebook
author: "Joe Edwards"
---

```{r}
library(vegan)
library(MASS)
library(tidyverse)
library(broom)
library(biobroom)
library(tidyMB)
library(RColorBrewer)
```

```{r}
## Do Not Run
#wExp <- readRDS("~/RMB/SoilDomestication/Data/wExp.rds")
nExp <- wExp %>% 
  filter(Site == "Arkansas") %>% 
  group_by(variable) %>% 
  filter(sum(value > 0) / n() > 0.05) %>% 
  ungroup() %>% 
  select_if(~!all(is.na(.)))

write_rds(nExp,"~/RMB/SoilDomestication/Data/native_plants_data.rds")


nExp <- readRDS("~/RMB/SoilDomestication/Data/native_plants_data.rds") %>% 
  filter(SampleID != "Arw.22") %>% 
  mutate(cpm = round(value * (1000000 / depth)))
tax <- readRDS("~/RMB/SoilDomestication/Data/gg_otus_tax.rds")
```

```{r}
raref <- function(x, sampling_depth, value = "value", otus = "variable") {
  rare_values <- data.frame(table(sample(x$`variable`, sampling_depth, replace = T, prob = x$`value`/x$depth)))
  names(rare_values) <- c(otus, "rare_value")
  return(suppressMessages(left_join(x, rare_values, by = otus) %>% replace_na(list(rare_value = 0))))
}
```


```{r} 
nPC <- tidy_pcoa(nExp %>% group_by(variable) %>% filter(sum(value) > 0) %>% mutate(logRA = log2(RA*1000 + 1)), dist = "bray", value = "logRA", keep_loadings = T)

nPC_plot <- nPC$axes %>% 
  ggplot(aes(MDS1, MDS2, color = host_common_name, shape = Compartment)) +
  geom_point(size = 3, alpha = 0.9) +
  #stat_ellipse(aes(group = Compartment), color = "black") +
  scale_color_brewer(palette = "Set1", direction = -1) +
  scale_shape_manual(values = c(1, 18, 16)) +
  labs(x = paste("PCo1 (", round(nPC$eigen_vals[1] * 100, 2), "%)", sep = ""), 
       y = paste("PCo2 (", round(nPC$eigen_vals[2] * 100, 2), "%)", sep = "")) +
  theme(text = element_text(size = 12),
        panel.background = element_rect(fill = "white"),
        panel.border = element_rect(fill = NA),
        axis.ticks.x = element_blank(),
        axis.ticks.y = element_blank())

nPC_plot
```
```{r}
nPC$axes %>% 
  dplyr::select(SampleID, host_common_name, Compartment, MDS1:MDS5) %>% 
  gather(key = Axis, value = value, -c(SampleID, host_common_name, Compartment)) %>% 
  group_by(Axis) %>% filter(Compartment != "Bulk Soil") %>% 
  nest() %>% 
  mutate(mod = map(data, ~aov(value ~ Compartment + host_common_name, .))) %>% 
  unnest(map(mod, ~tidy(.))) %>% 
  filter(term == "host_common_name")


nPC$axes %>% 
  dplyr::select(SampleID, host_common_name, Compartment, MDS1:MDS5) %>% 
  gather(key = Axis, value = value, -c(SampleID, host_common_name, Compartment)) %>% 
  group_by(Axis) %>% filter(Compartment != "Bulk Soil" & host_common_name != "Rice") %>% 
  nest() %>% 
  mutate(mod = map(data, ~aov(value ~ Compartment + host_common_name, .))) %>% 
  unnest(map(mod, ~tidy(.))) %>% 
  filter(term == "host_common_name")
```

```{r}
long_adonis(nExp %>% mutate(RA2 = log2((RA*1000) + 1)), value = "RA2", formula = "Compartment * host_common_name") %>% tidy()
```
```{r}
nExp %>% mutate(RA2 = log2((RA*1000) + 1)) %>% 
  group_by(Compartment) %>% 
  filter(Compartment != "Bulk Soil") %>% nest() %>% 
  mutate(adon = map(data, ~long_adonis(., value = "RA2", formula = "host_common_name"))) %>% 
  unnest(map(adon, ~tidy(.)))
```
```{r}
a_div <- nExp %>% 
  mutate(Compartment = fct_relevel(Compartment, "Bulk Soil", "Rhizosphere", "Endopshere")) %>% 
  mutate(host_common_name = relevel(factor(host_common_name), ref = "Rice")) %>% 
  group_by(Compartment, host_common_name, SampleID) %>% 
  filter(RA > 0) %>% 
  summarise(richness = sum(RA > 0), shannon = 2^(-sum(log2(RA) * RA))) 

a_div %>% 
  gather(key = metric, value = diversity, -c(Compartment, host_common_name, SampleID)) %>% 
  group_by(Compartment, metric) %>% 
  nest() %>% 
  filter(Compartment != "Bulk Soil") %>% 
  mutate(mod = map(data, ~aov(diversity ~ host_common_name, .))) %>% 
  unnest(map(mod, ~tidy(TukeyHSD(.))))

a_div %>% group_by(1) %>% nest() %>% 
  mutate(mod = map(data, ~aov(richness ~ Compartment * host_common_name, .))) %>% 
  unnest(map(mod, ~tidy(.)))

a_div %>% 
  ggplot(aes(Compartment, richness, color = host_common_name)) +
  geom_point(position = position_jitterdodge(jitter.width = 0.2, jitter.height = 0)) +
  scale_color_manual(values = c("#4DAF4A", "#FF7F00", "#984EA3", "#377EB8", "#E41A1C")) +
  theme_minimal() +
  labs(x = "", y = "Detectable OTUs") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1), legend.position = "none")
```

From the above PERMANOVA, host species does have a strong affect on the microbiome, but looking at the PCoA it is not clear whether the genotypic effect exists only because rice is so weird compared to the other plants.
```{r}
library(ape)
species_colors <- data.frame(host_common_name = c("Soil", "Rice", "Sedge", "Redstem", "Mudplantain"),
                             colorz = as.character(c("#E41A1C", "#377EB8", "#4DAF4A", "#984EA3", "#FF7F00")),
                             colorz2 = c("firebrick2", "palegreen4", "steelblue", "orchid4", "darkorange1"))
nDist <- tidyMB::wide_distance(nExp %>% mutate(RA2 = log2(RA* 1000 + 1)), value = "RA2")
nMD <- grab_metadata(nExp)
nHCL <- hclust(nDist, method = "average")
nMD <- nMD[match(nHCL$labels, nMD$SampleID),] %>% 
  inner_join(species_colors, by = "host_common_name")
nHCL$labels <- nMD$Compartment

plot(as.phylo(nHCL), lwd=2, tip.color = as.character(nMD$colorz2), use.edge.length = FALSE)
```
The tree makes it appear that there is significant clustering of the different species. Let's insepct this using a statistical approach. I will run a permanova on each compartment (excluding the bulk soil) after removing the rice samples.
```{r}
nExp %>% 
  filter(Compartment != "Bulk Soil" & host_common_name != "Rice") %>% 
  group_by(Compartment) %>% 
  mutate(RA2 = log2(RA*1000 + 1)) %>% 
  nest() %>% 
  mutate(permanova = map(data, ~long_adonis(., value = "RA2", formula = "host_common_name"))) %>% 
  unnest(map(permanova, ~tidy(.)))
```

```{r}
nExp %>% 
  mutate(RA2 = log2((RA*1000) + 1)) %>% 
  group_by(Compartment) %>% 
  nest() %>% 
  filter(Compartment != "Bulk Soil") %>% 
  mutate(ad = map(data, ~long_adonis(., value = "RA2", formula = "host_common_name"))) %>% 
  unnest(map(ad, ~tidy(.)))
```
Here we have the permanova results for how host species affects the microbiome within each compartment. The top table shows the results when rice is excluded from the analysis. The lower model includes rice. By comparing the R2 values, we can see that excluding rice from the analysis drastically reduces the effect size, but the result is still significant.

## Phyla Stuff
```{r}
phyla_abund <- nExp %>% 
  inner_join(tax, by = "variable") %>% 
  group_by(SampleID, Compartment, host_common_name, Phylum2, depth) %>% 
  summarise(phy_total = sum(value)) %>% 
  mutate(prop = (phy_total + 1) / (depth + 1))

phyla_abund %>% 
  group_by(Phylum2) %>% 
  nest() %>% 
  mutate(total = map_dbl(data, ~sum(.x$prop))) %>% 
  top_n(15, total) %>% 
  unnest(data) %>% 
  mutate(Compartment = fct_relevel(Compartment, "Bulk Soil", "Rhizosphere", "Endosphere")) %>% 
  group_by(Compartment, host_common_name, SampleID) %>% 
  nest() %>% 
  group_by(Compartment) %>% 
  arrange(host_common_name) %>% 
  mutate(order = 1:n()) %>% 
  unnest() %>% 
  ggplot(aes(order, prop * 100, fill = Phylum2)) +
  geom_bar(stat = "identity", width = 1) +
  geom_point(aes(x = order, y = -2, color = host_common_name), shape = 15) +
  facet_grid(.~Compartment, scales = "free_x") +
  scale_fill_manual(values = c(brewer.pal(11, "RdGy"),brewer.pal(5, "Blues")[-1])) +
  scale_color_brewer(palette = "Set1", direction = -1) +
  labs(x = "", y = "Percent of Reads") +
  theme_minimal() +
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        legend.position = "bottom",
        text = element_text(size = 12))
```

Let's use beta regression to find phyla whose relative abundance is signficantly different from rice. Betaregression requires that the dependent variable be in the [0,1] interval. If it's 0 or 1 it will not work. So the proportions that are being used here are calculated as (phylum counts within sample + 1) / (sequencing depth of sample + 1).
```{r}
library(betareg)
safe_betareg <- possibly(betareg, NA_real_)
nPhyModel_host <- phyla_abund %>% 
  ungroup() %>% 
  filter(Compartment != "Bulk Soil") %>% 
  mutate(host_common_name = relevel(factor(host_common_name), "Rice")) %>% 
  group_by(Phylum2, Compartment) %>% 
  nest() %>% 
  mutate(model = map(data, ~safe_betareg(prop ~ host_common_name, .))) %>% 
  unnest(map(model, ~tidy(.))) 

nPhyModel_host %>% 
  filter(term != "(Intercept)" & term != "(phi)") %>% 
  group_by(Compartment) %>% 
  mutate(p.adj = p.adjust(p.value, "BH")) %>% 
  filter(p.adj <= 0.05) %>% 
  mutate(direction = ifelse(estimate > 0, "Other", "Rice")) %>% 
  dplyr::count(direction, Compartment, Phylum2) %>% 
  filter(n == 3)

nPhyModel_host %>% 
  filter(term != "(Intercept)" & term != "(phi)") %>% 
  group_by(Compartment) %>% 
  mutate(p.adj = p.adjust(p.value, "BH")) %>% 
  filter(p.adj <= 0.05) %>% 
  mutate(direction = ifelse(estimate > 0, "Other", "Rice")) %>% 
  dplyr::count(direction, Compartment, Phylum2) %>% 
  filter(n == 3) %>% 
  dplyr::count(direction, Compartment)
```


```{r}
nPhyModel_compartment <- phyla_abund %>% 
  ungroup() %>% 
  mutate(Compartment = fct_relevel(factor(Compartment), "Bulk Soil", "Rhizosphere", "Endosphere")) %>% 
  group_by(Phylum2) %>% 
  nest() %>% 
  mutate(model = map(data, ~safe_betareg(prop ~ Compartment, .))) %>% 
  unnest(map(model, ~tidy(.))) 

nPhyModel_compartment %>% 
  filter(term != "(Intercept)" & term != "(phi)") %>% 
  mutate(term = gsub("Compartment", "", term)) %>% 
  mutate(p.adj = p.adjust(p.value, "BH")) %>% 
  inner_join(phyla_abund %>% group_by(Phylum2, Compartment) %>% summarise(mean_ab = mean(prop)), by = c(c("term" = "Compartment"), "Phylum2")) %>% 
  mutate(Compartment = fct_relevel(term, "Rhizosphere", "Endosphere")) %>%
  ggplot(aes(Compartment, estimate, group = Phylum2, color = ifelse(p.adj <= 0.05, Compartment, "ns"), shape = ifelse(p.adj <= 0.05, "sig", "ns"), size = log2(mean_ab * 1000))) +
  geom_line(color = 'black', alpha = 0.5, size = 0.5) +
  geom_point(alpha = 0.5) +
  scale_shape_manual(values = c(1, 16)) +
  scale_color_manual(values = c("#DFC27D", "#80CDC1", "black")) +
  #scale_color_manual(values = c("darkmagenta", "steelblue", "black")) +
  scale_size_continuous(range = c(0,10), breaks = c(-4, 0, 4, 8)) +
  theme_minimal() +
  labs(x = "") +
  theme(text = element_text(size = 12))

```


## Differential OTU Abundance
Let's find the OTUs that are different between the microbiomes of the different plants. In this analysis I will limit the contrasts to comparisons between the other plants and rice.
```{r}
## I commented out this whole chunk because I don't want to run it again. I saved the results as an rds object and will read it in for working in the future.

DESeq2_plant <- nExp %>% 
   filter(Compartment != "Bulk Soil") %>% 
   group_by(variable) %>% 
   filter(sum(value > 0) / n() > 0.1) %>% 
   mutate(host_common_name = relevel(factor(host_common_name), ref = "Rice")) %>% 
   mutate(group = paste(Compartment, host_common_name, sep = ".")) %>% 
   group_by(group_var = "DESeq2") %>% 
   nest() %>% 
   mutate(DGEL = map(data, ~tidyDGEL(., value = "value", group_column = "group", method = "DESeq2", formula = "~ group"))) %>% 
   mutate(dds = map(DGEL, ~DESeq(.))) %>% 
   mutate(Sedge_RS = map(dds, ~lfcShrink(., contrast = c("group", "Rhizosphere.Rice", "Rhizosphere.Sedge")))) %>% 
   mutate(Mudplantain_RS = map(dds, ~lfcShrink(., contrast = c("group", "Rhizosphere.Rice", "Rhizosphere.Mudplantain")))) %>% 
   mutate(Redstem_RS = map(dds, ~lfcShrink(., contrast = c("group", "Rhizosphere.Rice", "Rhizosphere.Redstem")))) %>% 
   mutate(Sedge_ES = map(dds, ~lfcShrink(., contrast = c("group", "Endosphere.Rice", "Endosphere.Sedge")))) %>% 
   mutate(Mudplantain_ES = map(dds, ~lfcShrink(., contrast = c("group", "Endosphere.Rice", "Endosphere.Mudplantain")))) %>% 
   mutate(Redstem_ES = map(dds, ~lfcShrink(., contrast = c("group", "Endosphere.Rice", "Endosphere.Redstem")))) %>% 
   dplyr::select(Mudplantain_RS, Sedge_RS, Redstem_RS, Mudplantain_ES, Sedge_ES, Redstem_ES, group_var) %>% 
   gather(key = taxon, value = results, -group_var)

saveRDS(DESeq2_plant, file = "~/RMB/SoilDomestication/Data/weeds_deseq.rds")
```

```{r}
DESeq2_plant <- read_rds("~/RMB/SoilDomestication/Data/weeds_deseq.rds")

np_MA_plot <- DESeq2_plant %>% 
  unnest(map(results, ~tidy(.))) %>% 
  separate(taxon, sep = "_", into = c("Taxon", "Compartment")) %>% 
  mutate(direction = ifelse(estimate < 0, "Weed", "Rice")) %>% 
  mutate(p.adjusted = ifelse(is.na(p.adjusted), 1, p.adjusted)) %>% 
  mutate(sig = ifelse(p.adjusted <= 0.05, "sig", "ns")) %>% 
  mutate(sig_direction = ifelse(p.adjusted <= 0.05, direction, NA)) %>% 
  mutate(color = ifelse(sig_direction == "Rice", "Rice", Taxon)) %>% 
  mutate(color = ifelse(is.na(color), "ns", color)) %>% 
  dplyr::add_count(Compartment, sig_direction, gene) %>% 
  mutate(cons = ifelse(p.adjusted <= 0.05 & n == 3, "cons", "nc")) %>% 
  mutate(Compartment = fct_recode(factor(Compartment, levels = c("RS", "ES")),
                                  "Rhizosphere" = "RS",
                                  "Endosphere" = "ES")) %>% 
  ggplot(aes(baseMean, estimate, color = color, alpha = sig, shape = cons)) +
  geom_point(size = 1) +
  facet_grid(Compartment ~ Taxon, scales = "free_y") +
  scale_x_log10() +
  scale_shape_manual(values = c(16, 1)) +
  scale_alpha_manual(values = c(0.1, 0.7)) +
  scale_color_manual(values = c("#FF7F00", "grey20", "#984EA3", "#4DAF4A", "#377EB8")) +
  theme_minimal()

np_MA_plot
```

Why do this analysis? From our data so far, it appears that rice is an outlier. There must be OTUs that set it apart from the other plants. There are likely OTUs that are commonly enriched in rice compared to the other plants and OTUs that are commonly enriched in the other plants compared to rice. Let's go ahead and make some venn diagrams to figure this out.
```{r}
library(VennDiagram)

deseq_results <- DESeq2_plant %>% 
  unnest(map(results, ~tidy(.))) %>% 
  separate(taxon, sep = "_", into = c("Taxon", "Compartment")) %>% 
  mutate(direction = ifelse(estimate < 0, "Weed", "Rice")) %>% 
  filter(p.adjusted <= 0.05) %>% 
  mutate(Compartment = gsub("ES", "Endosphere", Compartment),
         Compartment = gsub("RS", "Rhizosphere", Compartment)) %>% 
  dplyr::rename(variable = gene)

es_venn_rice <- list(Mudplantain = deseq_results %>% filter(Compartment == "Endosphere" & direction == "Rice" & Taxon == "Mudplantain") %>% pull(variable),
                     Sedge = deseq_results %>% filter(Compartment == "Endosphere" & direction == "Rice" & Taxon == "Sedge") %>% pull(variable),
                     Redstem = deseq_results %>% filter(Compartment == "Endosphere" & direction == "Rice" & Taxon == "Redstem") %>% pull(variable))
rs_venn_rice <- list(Mudplantain = deseq_results %>% filter(Compartment == "Rhizosphere" & direction == "Rice" & Taxon == "Mudplantain") %>% pull(variable),
                     Sedge = deseq_results %>% filter(Compartment == "Rhizosphere" & direction == "Rice" & Taxon == "Sedge") %>% pull(variable),
                     Redstem = deseq_results %>% filter(Compartment == "Rhizosphere" & direction == "Rice" & Taxon == "Redstem") %>% pull(variable))
es_venn_weed <- list(Mudplantain = deseq_results %>% filter(Compartment == "Endosphere" & direction == "Weed" & Taxon == "Mudplantain") %>% pull(variable),
                     Sedge = deseq_results %>% filter(Compartment == "Endosphere" & direction == "Weed" & Taxon == "Sedge") %>% pull(variable),
                     Redstem = deseq_results %>% filter(Compartment == "Endosphere" & direction == "Weed" & Taxon == "Redstem") %>% pull(variable))
rs_venn_weed <- list(Mudplantain = deseq_results %>% filter(Compartment == "Rhizosphere" & direction == "Weed" & Taxon == "Mudplantain") %>% pull(variable),
                     Sedge = deseq_results %>% filter(Compartment == "Rhizosphere" & direction == "Weed" & Taxon == "Sedge") %>% pull(variable),
                     Redstem = deseq_results %>% filter(Compartment == "Rhizosphere" & direction == "Weed" & Taxon == "Redstem") %>% pull(variable))

venn.diagram(es_venn_rice, filename = "~/RMB/SoilDomestication/Figures/Figure3/r_es_venn.svg", imagetype = "svg", height = 2, width = 2, euler.d = T, fill = c("darkorange1", "steelblue", "orchid4"), alpha = rep(0.3, 3), lwd = 0, category.names=c("", "", ""))
venn.diagram(rs_venn_rice, filename = "~/RMB/SoilDomestication/Figures/Figure3/r_rs_venn.svg", imagetype = "svg", height = 2, width = 2, euler.d = T, fill = c("darkorange1", "steelblue", "orchid4"), alpha = rep(0.3, 3), lwd = 0, category.names=c("", "", ""))
venn.diagram(es_venn_weed, filename = "~/RMB/SoilDomestication/Figures/Figure3/w_es_venn.svg", imagetype = "svg", height = 2, width = 2, euler.d = T, fill = c("darkorange1", "steelblue", "orchid4"), alpha = rep(0.3, 3), lwd = 0, category.names=c("", "", ""))
venn.diagram(rs_venn_weed, filename = "~/RMB/SoilDomestication/Figures/Figure3/w_rs_venn.svg", imagetype = "svg", height = 2, width = 2, euler.d = T, fill = c("darkorange1", "steelblue", "orchid4"), alpha = rep(0.3, 3), lwd = 0, category.names=c("", "", ""))
```

There are quite a few OTUs specifically enriched in rice or the weeds, let's see what phyla these belong to.
```{r}
intersect_deseq <- DESeq2_plant %>% 
  unnest(map(results, ~tidy(.))) %>% 
  separate(taxon, sep = "_", into = c("Taxon", "Compartment")) %>% 
  mutate(direction = ifelse(estimate < 0, "Weed", "Rice")) %>% 
  filter(p.adjusted <= 0.05) %>% 
  dplyr::count(direction, Compartment, gene) %>% 
  dplyr::add_count(direction, Compartment) %>% 
  filter(n == 3) %>% 
  mutate(Compartment = gsub("ES", "Endosphere", Compartment),
         Compartment = gsub("RS", "Rhizosphere", Compartment)) %>% 
  dplyr::rename(variable = gene)

intersect_deseq %>% 
  inner_join(tax, by = "variable") %>% 
  dplyr::count(Compartment, direction, Phylum2) %>% 
  mutate(Compartment  = fct_relevel(factor(Compartment), "Rhizosphere", "Endosphere")) %>% 
  mutate(nnn2 = ifelse(direction == "Rice", nnn, -nnn)) %>% 
  group_by(Compartment) %>% 
  arrange(nnn2) %>% 
  mutate(order = 1:n()) %>% 
  mutate(direction_multiplier = ifelse(nnn2 > 0, 1, -1)) %>% 
  mutate(hjust = ifelse(nnn2 > 0, 0, 1)) %>% 
  mutate(location = ifelse(nnn >= abs(10), (1)*direction_multiplier, (nnn+0.5) * direction_multiplier)) %>% 
  mutate(color = ifelse(nnn >= abs(10), "white", "black")) %>% 
  ggplot(aes(order, nnn2, lable = "Phylum2", fill = factor(direction_multiplier))) +
  geom_bar(stat = "identity") +
  geom_text(aes(x = order, y = location, label = Phylum2, color = color, hjust = hjust), size = 3.5) +
  scale_color_manual(values = c("black", "white")) +
  facet_wrap( ~ Compartment, scales = "free") +
  theme_minimal() +
  theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), legend.position = "none") +
  coord_flip() +
  scale_fill_manual(values = c("black", "#4DAF4A"))
```
```{r}
nUniverse <- nExp %>% 
  inner_join(tax, by = "variable") %>% 
  group_by(variable) %>% 
  filter(sum(value > 0) / n() > 0.1) %>% 
  dplyr::count(variable, Phylum2) %>% ungroup() %>% 
  dplyr::count(Phylum2) %>% dplyr::rename(universe_phy = nn) %>% 
  mutate(universe_size = sum(universe_phy))
  
intersect_deseq %>% 
  inner_join(tax, by = "variable") %>% 
  dplyr::count(Compartment, direction, Phylum2) %>% 
  dplyr::rename(group_n = nnn) %>% 
  ungroup() %>% 
  group_by(Phylum2) %>%  
  group_by(direction, Compartment) %>% 
  mutate(group_size = sum(group_n)) %>% 
  inner_join(nUniverse, by = "Phylum2") %>% 
  mutate(phyp = phyper(group_n, universe_phy, universe_size - group_size, group_size, lower.tail = F)) %>% 
  mutate(p.adj = p.adjust(phyp, "fdr")) %>% 
  filter(p.adj <= 0.05) %>% 
  filter(group_n > 2)
```


One of the prevailing phyla coming out as being specifically enriched in rice is Euryarachaeota, which contains methanogenic archaea. Let's go a bit deeper here.
```{r}
rice_specific_mathanogen_otus <- intersect_deseq %>% 
  inner_join(tax, by = "variable") %>% 
  filter(grepl("Methano", Family) & direction == "Rice") %>% 
  dplyr::select(Compartment, variable) %>% 
  inner_join(tax, by = "variable")

methanogen_data <- nExp %>% 
  inner_join(tax, by = "variable") %>% 
  filter(grepl("Methano", Family)) %>% 
  group_by(variable, Compartment) %>%
  filter(sum(value) > 0) %>% 
  summarise(mean_ab = mean(RA)) %>% 
  group_by(Compartment) %>% 
  arrange(-mean_ab) %>% 
  mutate(rank = 1:n())

mg_plot <- rbind(methanogen_data %>% filter(!variable%in%rice_specific_mathanogen_otus) %>%  mutate(color = "not_rice"),
                 methanogen_data %>% inner_join(rice_specific_mathanogen_otus, by = c("variable", "Compartment")) %>% mutate(color = "rice")) %>% 
  filter(Compartment != "Bulk Soil") %>% ungroup() %>% 
  mutate(Compartment = fct_relevel(factor(Compartment), "Rhizosphere", "Endosphere")) %>% 
  ggplot(aes(rank, mean_ab, color = color)) +
  geom_point(size = 2) +
  facet_grid(.~Compartment) +
  scale_y_log10() +
  scale_color_manual(values = c("black", "red")) +
  theme_minimal() +
  theme(text = element_text(size = 12))
```


What about methanotrophic eubacteria?
```{r}
methanotroph_otus <- intersect_deseq %>% 
  inner_join(tax, by = "variable") %>% 
  filter(grepl("Methyl", Family) & Genus != "Pleomorphomonas") 

methanotroph_data <- nExp %>% 
  inner_join(tax, by = "variable") %>% 
  filter(grepl("Methyl", Family) & Genus != "Pleomorphomonas") %>% 
  group_by(variable, Compartment) %>%
  filter(sum(value) > 0) %>% 
  summarise(mean_ab = mean(RA)) %>% 
  group_by(Compartment) %>% 
  arrange(-mean_ab) %>% 
  mutate(rank = 1:n())

mt_plot <- rbind(methanotroph_data %>% filter(!variable%in%unique(methanotroph_otus$variable)) %>%  mutate(color = "ns"),
                 methanotroph_data %>% inner_join(methanotroph_otus, by = c("variable", "Compartment")) %>% mutate(color = direction)) %>% 
  filter(Compartment != "Bulk Soil") %>% ungroup() %>% 
  mutate(Compartment = fct_relevel(factor(Compartment), "Rhizosphere", "Endosphere")) %>% 
  ggplot(aes(rank, mean_ab, color = color, shape = color)) +
  geom_point(size = 2) +
  facet_grid(.~Compartment) +
  scale_y_log10() +
  scale_shape_manual(values = c(16,1,16)) +
  scale_color_manual(values = c("black", "dodgerblue", "dodgerblue")) +
  theme_minimal() +
  theme(text = element_text(size = 12))
```
```{r}
library(gridExtra)
grid.arrange(mg_plot, mt_plot)
```
```{r}
mg_mt_ratios <- rbind(nExp %>% inner_join(tax, by = "variable") %>% filter(grepl("Methano", Family)) %>% mutate(classification = "Methanogen"),
                      nExp %>% inner_join(tax, by = "variable") %>% filter(grepl("Methyl", Family) & Genus != "Pleomorphomonas")%>% mutate(classification = "Methanotroph")) %>% 
  group_by(SampleID, Compartment, host_common_name, classification) %>% 
  summarise(total = sum(RA)) %>% 
  tidyr::spread(key = classification, value = total) %>% 
  mutate(ratio = Methanogen / Methanotroph)

mg_mt_ratios %>% 
  ungroup() %>% 
  filter(Compartment != "Bulk Soil") %>% 
  mutate(Compartment = fct_relevel(factor(Compartment), "Rhizosphere", "Endosphere")) %>% 
  ggplot(aes(Compartment, log2(ratio), color = host_common_name)) +
  geom_hline(yintercept = 0, linetype = "dashed") +
  geom_point(position = position_jitterdodge(jitter.width = 0.1, jitter.height = 0), size = 2) +
  scale_color_manual(values = c("#FF7F00", "#984EA3", "#4DAF4A", "#377EB8")) +
  theme_minimal() +
  theme(text = element_text(size = 12)) +
  labs(x = "", y = "log2(Methanogen / Methanotroph)")

mg_mt_ratios %>% 
  filter(Compartment != "Bulk Soil" & ratio > 0 ) %>% 
  group_by(Compartment) %>% 
  mutate(host_common_name = relevel(factor(host_common_name), ref = "Rice")) %>% 
  nest() %>% 
  mutate(model = map(data, ~lm(log2(ratio) ~ host_common_name, .))) %>% 
  unnest(map(model, ~tidy(.)))

mg_mt_ratios %>% 
  filter(Compartment != "Bulk Soil" & ratio > 0 ) %>% ungroup() %>% 
  mutate(host_common_name = relevel(factor(host_common_name), ref = "Rice")) %>% 
  group_by(1) %>% 
  nest() %>% 
  mutate(model = map(data, ~aov(log2(ratio) ~ Compartment*host_common_name, .))) %>% 
  unnest(map(model, ~tidy(TukeyHSD(.)))) %>% 
  filter(adj.p.value <= 0.05)

```


## Comparisons to bulk soil
Looking at the above PCoA, the communities in the rice rhizosphere appear to be remarkably similar to bulk soil. Let's go more into depth and ask whether this is statistically true.
```{r}
bs_dists <- tidyMB::long_distance(nExp) %>% 
  filter(Compartment.x == "Bulk Soil" | Compartment.y == "Bulk Soil") %>% 
  rowwise() %>% 
  mutate(bs_column = min(c(Compartment.x, Compartment.y))) %>% 
  mutate(compartment_column = max(c(Compartment.x, Compartment.y))) %>% 
  mutate(taxon_name = min(c(host_common_name.x, host_common_name.y))) %>% 
  ungroup() %>% 
  mutate(taxon_name = fct_relevel(factor(taxon_name), "Rice", "Mudplantain", "Redstem", "Sedge", "Soil")) %>% 
  mutate(compartment_column = fct_relevel(factor(compartment_column), "Bulk Soil", "Rhizosphere", "Endosphere"))

bs_dists %>% 
  ggplot(aes(compartment_column, value, color = taxon_name)) +
  geom_point(position = position_jitterdodge(jitter.width = 0.3), alpha = 0.8) + 
  scale_color_manual(values = c("#4DAF4A", "#FF7F00", "#984EA3", "#377EB8", "#E41A1C")) +
  theme_minimal() +
  labs(x = "", y = "Bray Dissimilarity") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1), text = element_text(size = 12))
```
It looks like rice has a reduced dissimilarity to bulk soil compared to the other plant species. Let's check whether this is significant
```{r}
bs_dists %>% 
  filter(compartment_column == "Rhizosphere") %>% 
  mutate(taxon_name = relevel(taxon_name, ref = "Rice")) %>% 
  aov(value ~ taxon_name, .) %>% 
  TukeyHSD(.) %>% 
  tidy(.)
```
From this ANOVA result, it looks like rice rhizosphere microbiomes are significantly more similar to the bulk soil microbiome than the other plant species.


```{r}
combo_pcoa <- bind_rows(nExp, mExp) %>% 
  ungroup() %>% 
  mutate(logRA = log2(RA*1000 + 1)) %>% ungroup() %>% 
  tidy_pcoa(., value = "logRA")

combo_pcoa$axes %>% 
  mutate(alph = ifelse(study == "domestication", 1, 0.1)) %>% 
  mutate(study = relevel(factor(study), ref = "domestication")) %>% 
  ggplot(aes(-MDS1, -MDS2, color = study, alpha = alph)) +
  geom_point() +
  theme_minimal() +
  scale_color_manual(values = c("black", brewer.pal(7, "Set2")[-4])) +
  labs(x = "MDS 1", y = "MDS 2") +
  guides(alpha = FALSE)
```

