---
title: "Native Plants Analysis"
output: html_notebook
---

```{r}
library(vegan)
library(broom)
library(MASS)
library(tidyverse)
library(broom)
library(biobroom)
library(tidyMB)
library(RColorBrewer)
```

```{r}
## Do Not Run
#wExp <- readRDS("~/RMB/SoilDomestication/Data/wExp.rds")
#nExp <- wExp %>% 
#  filter(Site == "Arkansas") %>% 
#  ungroup() %>% 
#  saveRDS("~/RMB/SoilDomestication/Data/native_plants_data.rds")

nExp <- readRDS("~/RMB/SoilDomestication/Data/native_plants_data.rds") %>% filter(SampleID != "Arw.22") %>% 
  mutate(cpm = round(value * (1000000 / depth)))
tax <- readRDS("~/RMB/SoilDomestication/Data/gg_otus_tax.rds")
```

```{r}
raref <- function(x, sampling_depth, value = "value", otus = "variable") {
  rare_values <- data.frame(table(sample(x$`variable`, sampling_depth, replace = T, prob = x$`value`/x$depth)))
  names(rare_values) <- c(otus, "rare_value")
  return(suppressMessages(left_join(x, rare_values, by = otus) %>% replace_na(list(rare_value = 0))))
}
```


```{r} 
nPC <- tidy_pcoa(nExp %>% group_by(variable) %>% filter(sum(value) > 0) %>% mutate(RA = RA*1000), dist = "bray", value = "RA", keep_loadings = T)

nPC$axes %>% 
  ggplot(aes(MDS1, MDS2, color = host_common_name)) +
  geom_point(size = 3, alpha = 0.7) +
  stat_ellipse(aes(group = Compartment), color = "black") +
  scale_color_brewer(palette = "Set1", direction = -1) +
  theme_minimal() +
  labs(x = paste("PCo1 (", round(nPC$eigen_vals[1] * 100, 2), "%)", sep = ""), 
       y = paste("PCo2 (", round(nPC$eigen_vals[2] * 100, 2), "%)", sep = "")) +
  theme(text = element_text(size = 12))
```

```{r}
long_adonis(nExp %>% mutate(RA2 = log2((RA*1000) + 1)), value = "RA2", formula = "Compartment * host_common_name") %>% tidy()
```

From the above PERMANOVA, host species does have a strong affect on the microbiome, but looking at the PCoA it is not clear whether the genotypic effect exists only because rice is so weird compared to the other plants.
```{r}
library(ape)
species_colors <- data.frame(host_common_name = c("Soil", "Rice", "Sedge", "Redstem", "Mudplantain"),
                             colorz = as.character(c("#E41A1C", "#377EB8", "#4DAF4A", "#984EA3", "#FF7F00")),
                             colorz2 = c("firebrick2", "palegreen4", "steelblue", "orchid4", "darkorange1"))
nDist <- tidyMB::wide_distance(nExp %>% mutate(RA2 = log2(RA* 1000 + 1)), value = "RA2")
nMD <- grab_metadata(nExp)
nHCL <- hclust(nDist, method = "average")
nMD <- nMD[match(nHCL$labels, nMD$SampleID),] %>% 
  inner_join(species_colors, by = "host_common_name")
nHCL$labels <- nMD$Compartment

plot(as.phylo(nHCL), lwd=2, tip.color = as.character(nMD$colorz2), use.edge.length = FALSE)
```
The tree makes it appear that there is significant clustering of the different species. Let's insepct this using a statistical approach. I will run a permanova on each compartment (excluding the bulk soil) after removing the rice samples.
```{r}
nExp %>% 
  filter(Compartment != "Bulk Soil" & host_common_name != "Rice") %>% 
  group_by(Compartment) %>% 
  mutate(RA2 = log2(RA*1000 + 1)) %>% 
  nest() %>% 
  mutate(permanova = map(data, ~long_adonis(., value = "RA2", formula = "host_common_name"))) %>% 
  unnest(map(permanova, ~tidy(.)))
```

```{r}
nExp %>% 
  mutate(RA2 = log2((RA*1000) + 1)) %>% 
  group_by(Compartment) %>% 
  nest() %>% 
  filter(Compartment != "Bulk Soil") %>% 
  mutate(ad = map(data, ~long_adonis(., value = "RA2", formula = "host_common_name"))) %>% 
  unnest(map(ad, ~tidy(.)))
```
Here we have the permanova results for how host species affects the microbiome within each compartment. The top table shows the results when rice is excluded from the analysis. The lower model includes rice. By comparing the R2 values, we can see that excluding rice from the analysis drastically reduces the effect size, but the result is still significant.

## Phyla Stuff
```{r}
library(MASS)
phyla_abund <- nExp %>% 
  inner_join(tax, by = "variable") %>% 
  group_by(SampleID, Compartment, host_common_name, Phylum2, depth) %>% 
  summarise(phy_total = sum(value)) %>% 
  mutate(prop = (phy_total + 1) / (depth + 1)) %>% 
  group_by(Compartment) %>% 
  mutate(alpha = fitdistr(prop, dbeta, start = list(shape1 = 1, shape2 = 10))$estimate[1],
         beta = fitdistr(prop, dbeta, start = list(shape1 = 1, shape2 = 10))$estimate[2]) %>% 
  mutate(emp_estimate = (phy_total + alpha + 1) / (depth + 1 + alpha + beta))

phyla_abund %>% 
  group_by(Phylum2) %>% 
  nest() %>% 
  mutate(total = map_dbl(data, ~sum(.x$prop))) %>% 
  top_n(15, total) %>% 
  unnest(data) %>% 
  mutate(Compartment = fct_relevel(Compartment, "Bulk Soil", "Rhizosphere", "Endosphere")) %>% 
  group_by(Compartment, host_common_name, SampleID) %>% 
  nest() %>% 
  group_by(Compartment) %>% 
  arrange(host_common_name) %>% 
  mutate(order = 1:n()) %>% 
  unnest() %>% 
  ggplot(aes(order, prop * 100, fill = Phylum2)) +
  geom_bar(stat = "identity", width = 1) +
  geom_point(aes(x = order, y = -2, color = host_common_name), shape = 15) +
  facet_grid(.~Compartment, scales = "free_x") +
  scale_fill_manual(values = c(brewer.pal(11, "RdGy"),brewer.pal(5, "Blues")[-1])) +
  scale_color_brewer(palette = "Set1", direction = -1) +
  labs(x = "", y = "Percent of Reads") +
  theme_minimal() +
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        legend.position = "bottom",
        text = element_text(size = 12))
```

```{r}
phy_host_lm <- phyla_abund %>% 
  group_by(Compartment, Phylum2) %>% 
  filter(host_common_name != "Soil") %>% 
  mutate(host_common_name = fct_relevel(host_common_name, "Rice", "Redstem", "Sedge", "Mudplantain")) %>% 
  nest() %>% 
  mutate(models = map(data, ~tidy(lm(log2(emp_estimate) ~ host_common_name, .)))) %>% 
  unnest(models) %>% 
  filter(term != "(Intercept)") %>% 
  mutate(term = gsub("host_common_name", "", term)) %>% 
  group_by(Compartment) %>% 
  mutate(p.adj = p.adjust(p.value, "fdr"))

phy_host_lm1 <- phyla_abund %>% 
  group_by(Compartment, Phylum2) %>% 
  filter(host_common_name != "Soil") %>% 
  mutate(host_common_name = fct_relevel(host_common_name, "Rice", "Redstem", "Sedge", "Mudplantain")) %>% 
  nest() %>% 
  mutate(models = map(data, ~tidy(lm(log2(phy_total + 1) ~ host_common_name, .)))) %>% 
  unnest(models) %>% 
  filter(term != "(Intercept)") %>% 
  mutate(term = gsub("host_common_name", "", term)) %>% 
  group_by(Compartment) %>% 
  mutate(p.adj = p.adjust(p.value, "fdr"))

phy_comp_lm <- phyla_abund %>% 
  group_by(Phylum2) %>% 
  nest() %>% 
  mutate(models = map(data, ~tidy(lm(log2(emp_estimate) ~ Compartment, .)))) %>% 
  unnest(models) %>% 
  filter(term != "(Intercept)") %>% 
  mutate(term = gsub("Compartment", "", term)) %>% 
  mutate(p.adj = p.adjust(p.value, "bon"))

phy_comp_lm %>% 
  filter(p.adj <= 0.05) %>% 
  mutate(Phylum2 = fct_reorder(factor(Phylum2), estimate)) %>% 
  ggplot(aes(Phylum2, estimate)) +
  geom_point() +
  facet_grid(. ~ term)

phy_comp_lm %>% 
  inner_join(phyla_abund %>% group_by(Phylum2, Compartment) %>% summarise(mean_ab = mean(emp_estimate)), by = c(c("term" = "Compartment"), "Phylum2")) %>% 
  mutate(Compartment = fct_relevel(term, "Rhizosphere", "Endosphere")) %>%
  ggplot(aes(Compartment, estimate, group = Phylum2, color = ifelse(p.adj <= 0.05, Compartment, "ns"), shape = ifelse(p.adj <= 0.05, "sig", "ns"), size = log2(mean_ab * 1000))) +
  geom_line(color = 'black', alpha = 0.5, size = 0.5) +
  geom_point(alpha = 0.5) +
  scale_shape_manual(values = c(1, 16)) +
  scale_color_manual(values = c("darkmagenta", "steelblue", "black")) +
  scale_size_continuous(range = c(0,10)) +
  theme_minimal()

phy_host_lm %>% 
  filter(p.adj <= 0.05) %>% ungroup() %>% 
  mutate(Compartment = fct_relevel(Compartment, "Rhizosphere", "Endosphere")) %>% 
  ggplot(aes(Phylum2, term, fill = estimate)) +
  geom_tile() +
  facet_grid(Compartment~.) +
  scale_fill_gradient2(low = "gold", high = "dodgerblue", mid = 'black') +
  theme(axis.text.x = element_text(angle = 30, hjust = 1))
  

```


## Differential OTU Abundance
```{r}
DESeq2_plant <- nExp %>% 
  filter(Compartment != "Bulk Soil") %>% 
  group_by(variable) %>% 
  filter(sum(value > 0) / n() > 0.1) %>% 
  mutate(host_common_name = relevel(factor(host_common_name), ref = "Rice")) %>% 
  mutate(group = paste(Compartment, host_common_name, sep = ".")) %>% 
  group_by(group_var = "DESeq2") %>% 
  nest() %>% 
  mutate(DGEL = map(data, ~tidyDGEL(., value = "value", group_column = "group", method = "DESeq2", formula = "~ group"))) %>% 
  mutate(dds = map(DGEL, ~DESeq(.))) %>% 
  mutate(Sedge_RS = map(dds, ~results(., contrast = c("group", "Rhizosphere.Rice", "Rhizosphere.Sedge")))) %>% 
  mutate(Mudplantain_RS = map(dds, ~results(., contrast = c("group", "Rhizosphere.Rice", "Rhizosphere.Mudplantain")))) %>% 
  mutate(Redstem_RS = map(dds, ~results(., contrast = c("group", "Rhizosphere.Rice", "Rhizosphere.Redstem")))) %>% 
  mutate(Sedge_ES = map(dds, ~results(., contrast = c("group", "Endosphere.Rice", "Endosphere.Sedge")))) %>% 
  mutate(Mudplantain_ES = map(dds, ~results(., contrast = c("group", "Endosphere.Rice", "Endosphere.Mudplantain")))) %>% 
  mutate(Redstem_ES = map(dds, ~results(., contrast = c("group", "Endosphere.Rice", "Endosphere.Redstem")))) %>% 
  dplyr::select(Mudplantain_RS, Sedge_RS, Redstem_RS, Mudplantain_ES, Sedge_ES, Redstem_ES, group_var) %>% 
  gather(key = taxon, value = results, -group_var)

saveRDS(DESeq2_plant, file = "~/RMB/SoilDomestication/Data/weeds_deseq.rds")
```

```{r}
DESeq2_plant <- read_rds("~/RMB/SoilDomestication/Data/weeds_deseq.rds")

DESeq2_plant %>% 
  unnest(map(results, ~tidy(.))) %>% 
  separate(taxon, sep = "_", into = c("Taxon", "Compartment")) %>% 
  mutate(direction = ifelse(estimate < 0, "Weed", "Rice")) %>% 
  mutate(p.adjusted = ifelse(is.na(p.adjusted), 1, p.adjusted)) %>% 
  mutate(sig = ifelse(p.adjusted <= 0.05, "sig", "ns")) %>% 
  mutate(sig_direction = ifelse(p.adjusted <= 0.05, direction, NA)) %>% 
  mutate(color = ifelse(sig_direction == "Rice", "Rice", Taxon)) %>% 
  mutate(color = ifelse(is.na(color), "ns", color)) %>% 
  dplyr::add_count(Compartment, sig_direction, gene) %>% 
  mutate(cons = ifelse(p.adjusted <= 0.05 & n == 3, "cons", "nc")) %>% 
  filter(abs(estimate) < 20) %>% 
  mutate(Compartment = fct_recode(factor(Compartment, levels = c("RS", "ES")),
                                  "Rhizosphere" = "RS",
                                  "Endosphere" = "ES")) %>% 
  ggplot(aes(baseMean, estimate, color = color, alpha = sig, shape = cons)) +
  geom_point(size = 1) +
  facet_grid(Compartment ~ Taxon, scales = "free_y") +
  scale_x_log10() +
  scale_shape_manual(values = c(16, 1)) +
  scale_alpha_manual(values = c(0.1, 0.7)) +
  scale_color_manual(values = c("#FF7F00", "grey20", "#984EA3", "#4DAF4A", "#377EB8")) +
  theme_minimal()
```

Why do this analysis? From our data so far, it appears that rice is an outlier. There must be OTUs that set it apart from the other plants. There are likely OTUs that are commonly enriched in rice compared to the other plants and OTUs that are commonly enriched in the other plants compared to rice. Let's go ahead and make some venn diagrams to figure this out.
```{r}
library(VennDiagram)

deseq_results <- DESeq2_plant %>% 
  unnest(map(results, ~tidy(.))) %>% 
  separate(taxon, sep = "_", into = c("Taxon", "Compartment")) %>% 
  mutate(direction = ifelse(estimate < 0, "Weed", "Rice")) %>% 
  filter(p.adjusted <= 0.05) %>% 
  mutate(Compartment = gsub("ES", "Endosphere", Compartment),
         Compartment = gsub("RS", "Rhizosphere", Compartment)) %>% 
  dplyr::rename(variable = gene)

es_venn_rice <- list(Mudplantain = deseq_results %>% filter(Compartment == "Endosphere" & direction == "Rice" & Taxon == "Mudplantain") %>% pull(variable),
                     Sedge = deseq_results %>% filter(Compartment == "Endosphere" & direction == "Rice" & Taxon == "Sedge") %>% pull(variable),
                     Redstem = deseq_results %>% filter(Compartment == "Endosphere" & direction == "Rice" & Taxon == "Redstem") %>% pull(variable))
rs_venn_rice <- list(Mudplantain = deseq_results %>% filter(Compartment == "Rhizosphere" & direction == "Rice" & Taxon == "Mudplantain") %>% pull(variable),
                     Sedge = deseq_results %>% filter(Compartment == "Rhizosphere" & direction == "Rice" & Taxon == "Sedge") %>% pull(variable),
                     Redstem = deseq_results %>% filter(Compartment == "Rhizosphere" & direction == "Rice" & Taxon == "Redstem") %>% pull(variable))
es_venn_weed <- list(Mudplantain = deseq_results %>% filter(Compartment == "Endosphere" & direction == "Weed" & Taxon == "Mudplantain") %>% pull(variable),
                     Sedge = deseq_results %>% filter(Compartment == "Endosphere" & direction == "Weed" & Taxon == "Sedge") %>% pull(variable),
                     Redstem = deseq_results %>% filter(Compartment == "Endosphere" & direction == "Weed" & Taxon == "Redstem") %>% pull(variable))
rs_venn_weed <- list(Mudplantain = deseq_results %>% filter(Compartment == "Rhizosphere" & direction == "Weed" & Taxon == "Mudplantain") %>% pull(variable),
                     Sedge = deseq_results %>% filter(Compartment == "Rhizosphere" & direction == "Weed" & Taxon == "Sedge") %>% pull(variable),
                     Redstem = deseq_results %>% filter(Compartment == "Rhizosphere" & direction == "Weed" & Taxon == "Redstem") %>% pull(variable))

venn.diagram(es_venn_rice, filename = "r_es_venn.svg", imagetype = "svg", height = 2, width = 2, euler.d = T, fill = c("darkorange1", "steelblue", "orchid4"), alpha = rep(0.3, 3), lwd = 0, category.names=c("", "", ""))
venn.diagram(rs_venn_rice, filename = "r_rs_venn.svg", imagetype = "svg", height = 2, width = 2, euler.d = T, fill = c("darkorange1", "steelblue", "orchid4"), alpha = rep(0.3, 3), lwd = 0, category.names=c("", "", ""))
venn.diagram(es_venn_weed, filename = "w_es_venn.svg", imagetype = "svg", height = 2, width = 2, euler.d = T, fill = c("darkorange1", "steelblue", "orchid4"), alpha = rep(0.3, 3), lwd = 0, category.names=c("", "", ""))
venn.diagram(rs_venn_weed, filename = "w_rs_venn.svg", imagetype = "svg", height = 2, width = 2, euler.d = T, fill = c("darkorange1", "steelblue", "orchid4"), alpha = rep(0.3, 3), lwd = 0, category.names=c("", "", ""))
```


```{r}
intersect_deseq <- DESeq2_plant %>% 
  unnest(map(results, ~tidy(.))) %>% 
  separate(taxon, sep = "_", into = c("Taxon", "Compartment")) %>% 
  mutate(direction = ifelse(estimate < 0, "Weed", "Rice")) %>% 
  filter(p.adjusted <= 0.05) %>% 
  dplyr::count(direction, Compartment, gene) %>% 
  dplyr::add_count(direction, Compartment) %>% 
  filter(n == 3) %>% 
  mutate(Compartment = gsub("ES", "Endosphere", Compartment),
         Compartment = gsub("RS", "Rhizosphere", Compartment)) %>% 
  dplyr::rename(variable = gene)

intersect_deseq %>% 
  inner_join(tax, by = "variable") %>% 
  dplyr::count(Compartment, direction, Phylum2) %>% 
  mutate(nnn2 = ifelse(direction == "Rice", nnn, -nnn)) %>% 
  group_by(Phylum2, Compartment) %>% 
  mutate(order_thing = sum(nnn)) %>% 
  filter(nnn > 1) %>% 
  ggplot(aes(fct_reorder(Phylum2, -order_thing), nnn2)) +
  geom_bar(stat = "identity") +
  geom_hline(yintercept = 0, color = "white") +
  facet_grid(. ~ Compartment) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 30, hjust = 1))

intersect_deseq %>% 
  inner_join(tax, by = "variable") %>% 
  dplyr::count(Compartment, direction, Phylum2) %>% 
  mutate(Compartment  = fct_relevel(factor(Compartment), "Rhizosphere", "Endosphere")) %>% 
  group_by(Compartment, direction) %>% 
  arrange(nnn) %>% 
  mutate(order = 1:n()) %>% 
  filter(nnn > 2) %>% 
  mutate(location = ifelse(nnn >= 22, nnn-10, nnn+0.5)) %>% 
  mutate(color = ifelse(nnn >= 22, "white", "black")) %>% 
  ggplot(aes(order, nnn, lable = "Phylum2")) +
  geom_bar(stat = "identity") +
  geom_text(aes(x = order, y = location, label = Phylum2, color = color), hjust = 0, size = 3.5) +
  scale_color_manual(values = c("black", "white")) +
  facet_wrap( ~ Compartment + direction, scales = "free") +
  theme_minimal() +
  theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), legend.position = "none") +
  coord_flip() +
  ylim(0, 35)
```

## Comparisons to bulk soil
Looking at the above PCoA, the communities in the rice rhizosphere appear to be remarkably similar to bulk soil. Let's go more into depth and ask whether this is statistically true.
```{r}
bs_dists <- tidyMB::long_distance(nExp) %>% 
  filter(Compartment.x == "Bulk Soil" | Compartment.y == "Bulk Soil") %>% 
  rowwise() %>% 
  mutate(bs_column = min(c(Compartment.x, Compartment.y))) %>% 
  mutate(compartment_column = max(c(Compartment.x, Compartment.y))) %>% 
  mutate(taxon_name = min(c(host_common_name.x, host_common_name.y))) %>% 
  ungroup() %>% 
  mutate(taxon_name = fct_relevel(factor(taxon_name), "Rice", "Mudplantain", "Redstem", "Sedge", "Soil")) %>% 
  mutate(compartment_column = fct_relevel(factor(compartment_column), "Bulk Soil", "Rhizosphere", "Endosphere"))

bs_dists %>% 
  ggplot(aes(compartment_column, value, color = taxon_name)) +
  geom_point(position = position_jitterdodge(jitter.width = 0.3), alpha = 0.8) + 
  scale_color_manual(values = c("#4DAF4A", "#FF7F00", "#984EA3", "#377EB8", "#E41A1C")) +
  theme_minimal() +
  labs(x = "", y = "Bray Dissimilarity") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1), text = element_text(size = 12))
```
It looks like rice has a reduced dissimilarity to bulk soil compared to the other plant species. Let's check whether this is significant
```{r}
bs_dists %>% 
  filter(compartment_column == "Rhizosphere") %>% 
  mutate(taxon_name = relevel(taxon_name, ref = "Rice")) %>% 
  aov(value ~ taxon_name, .) %>% 
  TukeyHSD(.) %>% 
  tidy(.)
```
From this ANOVA result, it looks like rice rhizosphere microbiomes are significantly more similar to the bulk soil microbiome than the other plant species.


