---
title: "Soil Domestication"
output: html_notebook
author: "Joe Edwards"
---
## Introductions
Here in this experiment, we have grown rice plants in  soils from rice fields and from locations that are adjacent to rice fields, but are not growing crops. The goal was to see if there are consistent differences between the microbiomes that rice plants assemble from these soils.

## Load in the libraries
Note that this analysis relies on the tidyMB package. This can be found on my github profile.
```{r}
library(vegan)
library(broom)
library(tidyverse)
library(broom)
library(biobroom)
library(tidyMB)
library(DESeq2)
```

## Load in the data
The portion that is commented out was when I was originally loading in the data as a subset of a much larger dataset. The wExp object was too big for me to post on github, so I trimmed it down. Please note that if you are running this analysis on your computer, you will need to change the PATHs to reflect where you have clones this github repo to.
```{r}
## Do Not Run
#wExp <- read_rds("~/Google Drive/RMB/Analyses/WEEDS/Manuscript/wExp.rds")
#wExp %>% 
#  filter(Experiment == "Mix" | Experiment == "Transplant") %>% 
#  ungroup() %>% 
#  group_by(variable) %>% 
#  filter(sum(value) > 0) %>% 
#  saveRDS("~/RMB/SoilDomestication/Data/dom_data.rds")

dExp <- readRDS("~/RMB/SoilDomestication/Data/dom_data.rds")
tax <- readRDS("~/RMB/SoilDomestication/Data/gg_otus_tax.rds")
```

## Betadiversity
First things first. What is the underlying structure of the data?
```{r}
dPC <- tidy_pcoa(dExp %>% mutate(RA = RA*1000), dist = "bray", value = "RA", keep_loadings = F)

dPC$axes %>% 
  ggplot(aes(MDS1, MDS2, color = From, shape = Experiment)) +
  geom_point(size = 3, alpha = 0.7) +
  scale_color_manual(values = c("dodgerblue", "gold")) +
  theme_minimal() +
  labs(x = paste("PCo1 (", round(nPC$eigen_vals[1] * 100, 2), "%)", sep = ""), y = paste("PCo2 (", round(nPC$eigen_vals[2] * 100, 2), "%)", sep = ""))
```

## Comparing rhizosphere to bulk soil distances
Remember that result where the rice rhizosphere is significantly more similar to the bulk soil microbiome than the rhizosphere microbiomes from the native plants? One of the hypotheses that came from that was that rice may be "domesticating" the soil microbiome. If you go to rice field, you will quickly realize that those plants are growing so closely together that, at some point in the season, there all of the roots will overlap and the top layer of soil is only composed of rhizosphere. We can test this hypothesis by growing rice plants in soil from rice fields and soil from uncultivated sites. In order to support this hypothesis that rice is "domesticating" field soil, we need to show a few things. One is that the distance between rhizospehre and bulk soil is greater in the "wild" soils. The other is that the rhizospheres of rice plants growing in wild and domesticated soils are mmore similar than the bulk soils.
```{r}
d_dist <- dExp %>% 
  long_distance(.)

rs_vs_bs <- d_dist %>% 
  filter(Compartment.x%in%c("Rhizosphere", "Bulk Soil")) %>% 
  filter(Compartment.y%in%c("Rhizosphere", "Bulk Soil")) %>% 
  filter(Compartment.x != Compartment.y) %>% 
  filter(From.x == From.y) %>% 
  filter(Experiment.x == Experiment.y) %>% 
  mutate(From = ifelse(From.x == "A", "Agricultural", "Non-agricultural"))

rs_vs_bs_plot <- rs_vs_bs %>% 
  ggplot(aes(From, value, fill = From)) +
  geom_jitter(width = 0.1, height = 0, alpha = 0.5) +
  geom_boxplot(alpha = 0.3, width = 0.5) +
  scale_fill_manual(values = c("dodgerblue", "gold")) +
  scale_shape_manual(values = c(1,4)) +
  theme_minimal() +
  labs(x = "", y = "Bray Dissimilarity") +
  theme(text = element_text(size =12), legend.position = "none", axis.text.x = element_text(angle = 30, hjust = 1))

rs_vs_bs_plot
```

The above plot appears to support my suspicion that rice plants growing in agricultural soil host a rhizosphere microbiome that is more similar to their corresponding bulk soils. Let's do some statistics to make sure these differences are "signficant".
```{r}
tidy(lm(value ~ From.x, data = rs_vs_bs))

rs_vs_bs %>% 
  group_by(Experiment.x) %>% 
  nest() %>% 
  mutate(mod = map(data, ~lm(value ~ From.x, .))) %>% 
  unnest(map(mod, ~tidy(.)))
```

Cool Cool. Looks lik this trend is significant. Let's compare the rhizospheres to see if they are more similar.

```{r}
same_comps <- d_dist %>% 
  filter(Compartment.x%in%c("Rhizosphere", "Bulk Soil")) %>% 
  filter(Compartment.y%in%c("Rhizosphere", "Bulk Soil")) %>% 
  filter(Compartment.x == Compartment.y) %>% 
  filter(Experiment.x == Experiment.y) %>% 
  mutate(type = ifelse(From.y == From.x, "aSelf", "bOut")) 

same_comps_plot <- same_comps %>% 
  ggplot(aes(Compartment.x, value)) +
  geom_jitter(width = 0.1, alpha = 0.7) +
  geom_boxplot(alpha = 0, width = 0.7) +
  facet_wrap(~type) +
  scale_shape_manual(values = c(1,4)) +
  theme_minimal() +
  labs(x = "", y = "Bray Dissimilarity") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1), text = element_text(size = 12))

same_comps_plot
```

```{r}
same_comps %>% 
  group_by(type) %>% 
  nest() %>% 
  mutate(mod = map(data, ~lm(value ~ Compartment.x, .))) %>% 
  unnest(map(mod, ~tidy(.)))
```
```{r}
grid.arrange(rs_vs_bs_plot, same_comps_plot, ncol = 2)
```


## Differentially abundant OTUs
We see relatively large differences between "domesticated" and "wild" soils. Let's find the OTUs that explain these differences.
```{r}
dom_deseq <- dExp %>% 
  group_by(variable) %>% 
  filter(sum(value > 0) / n() > 0.1) %>% 
  mutate(group = paste(Compartment, From, sep = ".")) %>% 
  group_by(group_var = "DESeq2", Experiment) %>% 
  nest() %>% 
  mutate(DGEL = map(data, ~suppressMessages(tidyDGEL(., value = "value", group_column = "group", method = "DESeq2", formula = "~ group")))) %>% 
  mutate(dds = map(DGEL, ~suppressMessages(DESeq(.)))) %>% 
  mutate(BulkSoil = map(dds, ~lfcShrink(., contrast = c("group", "Bulk Soil.A", "Bulk Soil.N")))) %>% 
  mutate(Rhizosphere = map(dds, ~lfcShrink(., contrast = c("group", "Rhizosphere.A", "Rhizosphere.N")))) %>% 
  mutate(Endosphere = map(dds, ~lfcShrink(., contrast = c("group", "Endosphere.A", "Endosphere.N")))) %>% 
  dplyr::select(BulkSoil, Rhizosphere, Endosphere, Experiment) %>% 
  gather(key = Compartment, value = results, -Experiment)

dom_deseq %>% 
  ungroup() %>% 
  write_rds(path = "~/RMB/SoilDomestication/Data/domestication_deseq.rds")
```

```{r}
dom_otus <- dom_deseq %>% 
  unnest(map(results, ~tidy(.))) %>% 
  group_by(Experiment, Compartment) %>% 
  mutate(direction = ifelse(estimate > 0, "A", "N")) %>% 
  filter(p.adjusted <= 0.05) %>% ungroup() %>% 
  dplyr::count(direction, Compartment, gene) %>% 
  filter(n == 2) %>% 
  dplyr::count(direction, gene, Compartment) %>% 
  dplyr::rename(variable = gene)

np_deseq <- read_rds("~/RMB/SoilDomestication/Data/weeds_deseq.rds")
rice_otus <- np_deseq %>% 
  unnest(map(results, ~tidy(.))) %>% 
  separate(taxon, sep = "_", into = c("Taxon", "Compartment")) %>% 
  mutate(direction = ifelse(estimate < 0, "Weed", "Rice")) %>% 
  filter(p.adjusted <= 0.05) %>% 
  dplyr::count(direction, Compartment, gene) %>% 
  dplyr::add_count(direction, Compartment) %>% 
  filter(n == 3) %>% 
  mutate(Compartment = gsub("ES", "Endosphere", Compartment),
         Compartment = gsub("RS", "Rhizosphere", Compartment)) %>% 
  dplyr::rename(variable = gene) %>% 
  filter(direction == "Rice")
  
```

### What is the overlap of rice enriched microbes with the domesticated microbes?
```{r}
nA <- length(unique(rice_otus$variable))
nB <- length(unique(dom_otus %>% filter(direction == "A") %>% pull(variable)))
nC <- length(union(nExp %>% group_by(variable) %>% filter(sum(value) > 0) %>% dplyr::count(variable) %>% pull(variable),
          dExp %>% group_by(variable) %>% filter(sum(value) > 0) %>% dplyr::count(variable) %>% pull(variable)))
nAB <- length(intersect(dom_otus %>% filter(direction == "A") %>% pull(variable), rice_otus$variable))

prob <- phyper(nAB, nB, nC-nB, nA, lower.tail = F)
```
There are `r nA` rice enriched microbes (compared to the weeds) and `r nB` OTUs that are enriched at least in one compartment of the plants growing in domesticated soils. Between the rice vs. weeds dataset and the soil domestication dataset, there are `r nC` OTUs. There are `r nAB` OTUs shared between the rice enriched and domesticated microbes. What is the probability that we whitness and overlap of `r nAB` OTUs given these parameters? We can treat this like the trusty old "black and white balls in an urn" analogy. If we have an urn filled with `r nC`, and `r nB` of these balls are black and the rest are white, what is the probability that if we draw `r nA` balls from the urn that `r nAB` of the balls will be black? We can use the hypergeometric function to figure out. After doing this calculation, we find that the probability of having an overlap of `r nAB` between the datasets is `r prob`.

We have this nice overlap. Let's show it in tree form.
```{r}
library(phyloseq)
gg_tree <- readRDS("~/RMB/SoilDomestication/Data/gg_tree.rds")



domMD <- data.frame(Compartment = c("aShared", rep(c("bBulkSoil", "cRhizosphere", "dEndosphere"), each = 2)),
                      Type = c("aShared",rep(c("Domesticated", "Wild"), 3)),
                      SampleID = paste("D", 0:6, sep = ""),
                    row.names = paste("D", 0:6, sep = ""))

dT <- dom_otus %>% 
  mutate(Type = ifelse(direction == "A", "Domesticated", "Wild")) %>% 
  mutate(Compartment = gsub("BulkSoil", "bBulkSoil", Compartment),
         Compartment = gsub("Rhizosphere", "cRhizosphere", Compartment),
         Compartment = gsub("Endosphere", "dEndosphere", Compartment)) %>% 
  bind_rows(rice_otus %>% filter(variable%in%intersect(dom_otus %>% filter(direction == "A") %>% pull(variable), rice_otus$variable)) %>% 
              dplyr::count(variable) %>% 
              mutate(Type = "aShared", Compartment = "aShared")) %>% 
  mutate(value = 1) %>% 
  inner_join(domMD, by = c("Compartment", "Type")) %>%
  dplyr::select(SampleID, variable, value) %>% 
  spread(SampleID, value, fill = 0) %>% 
  as.data.frame()
row.names(dT) <- dT$variable
dT$variable <- NULL  

dTax <- dom_otus %>% 
  inner_join(tax, by = "variable") %>% 
  spread(Compartment, direction) %>% 
  mutate(Phylum2 = as.character(Phylum2)) %>% 
  dplyr::rename(Domain = Kingdom) %>% 
  as.data.frame()
row.names(dTax) <- dTax$variable
dTax <- dTax[match(row.names(dT), row.names(dTax)),]


dom.physeq_sampleType <- phyloseq(otu_table(dT, taxa_are_rows = T), 
                       sample_data(domMD),
                       dTax,
                       gg_tree)
# Plot tree with metadata
plot_tree(dom.physeq_sampleType, color = "Type", ladderize = "left", shape = "Compartment", justify = "left", size = "Type") +
  scale_color_manual(values = c('black', "dodgerblue", "goldenrod")) +
  scale_shape_manual(values = c(1, 16, 15, 17)) +
  scale_size_manual(values = c(1,2,2,2))

# Plot tree with taxonomic information
dom.physeq_Phy <- phyloseq(otu_table(data.frame(J1 = rep(1, nrow(dT)), row.names = row.names(dT)), taxa_are_rows = T), 
                       tax_table(as.matrix(dTax)),
                       gg_tree)

plot_tree(dom.physeq_Phy, color = "Phylum2", ladderize = "left") 
```

Do we find some phyla belonging to the categories more than expected? That is, for instance, do we find that deltoproteobacteria are enriched more in domesticated soils than we would expect by random chance?
```{r}
dom_otus %>% 
  mutate(universe = nrow(.)) %>% 
  inner_join(tax, by = "variable") %>% 
  add_count(Phylum2) %>% dplyr::rename(phy_universe = n) %>% 
  add_count(direction) %>% dplyr::rename(group_size = n) %>% 
  dplyr::count(universe, phy_universe, group_size, direction, Phylum2) %>% 
  mutate(phyp = phyper(n, group_size, universe - group_size, phy_universe, lower.tail = F)) %>% 
  filter(n > 3) %>% filter(phyp <= 0.05) %>% arrange(direction, phyp)
```

