---
title: "Seedling Vigor Microbiome Analysis"
output: html_notebook
---
## Introduction

In our previous results, we show that soil microbiomes differ between rice field soils (domesticated) and soils from nearby non-agricultural sites (wild soils). This is perhaps not suprising. Rice fields receive added nutrients and are flooded throughout the season. Why would this *not* affect the community of microbes living in the soil? An interesting correlation, however, is that many of the microbes enriched in the domesticated soils are also found to be specifically in rice plants compared to other native plant species growing in rice fields. Therefore it appears that rice plants themselves are shifting the soil microbiome to reflect the microbiota they actually acquire.

One could imagine many possibilities for why this phenomenon may be beneficial or detrimental to the host rice plants. The experiment I will describe below was initially designed to address a hypothesis that being surrounded by a pre-assembled microbiome may relieve some of the pressure or energy allocation for acquiring a microbiome and allow the plant to reallocate energy into general growth. Therefore, we would expect that seeds germinated in domesticated soils would portray greater seedling vigor traits that seeds germinated in wild soils. This was just one hypothesis. I'm sure you can imagine many more...

If we want to test this hypothesis, we have to be crafty with the expeirment. We cannot just put plants in soil from the field and measure seedling vigor traits. These soils will differ for many characteristics (e.g. nutrient content, structure, etc.), so we needed a way of evening out the playing field between each tested soil. We decided to try a method where we inoculate a dilute amount of soil into a common matrix, germinate our seedlings, and measure traits and microbiomes. It was a bit more complex than that, so let me elaborate:

1) Collect soils. Soils were dry at the time of collection.
2) Moisten soils to reactivate microbes. Allow to incubate in the dark at 30 degrees C for two weeks.
3) Autoclave the common soil substrate. In this case we used UC Mix, which is commonly used to grow rice plants in the greenhouse. There were 2 rounds of autoclaving.
4) After the soil inoculum is ready, we inoculated the soil slurry at 1% into the UC Mix. I will detail this procedure more.
* We used 1.8 kg of UC Mix and 18 g of the soil inoculum.
* The soil inoculum was weighed and put into 1 L of 0.5 strength MS.
* We also had another treatment of sterilized inoculum. This was done by plant 18 g of the soil into 1L of 0.5 strength MS and autoclaving the slurry.
* The slurries were poured into 1.8 kg of the UC Mix and mixed by hand. I put gloves on each hand, sprayed with 70% ethanol, and allowed to dry.
* After mixing, lids were placed on the containers and aluminum foil was wrapped around the containers to keep them dark. They were then transported to the greenhouse and allowed to incubate for 2 more weeks.
5) After two weeks, soils were loaded into autoclaved P1000 tip boxes in equal amounts. Each tipbox could hold 800 grams of soil. Holes in the tip boxes were previously made using a soldering iron to allow for water to enter each well. There were 8 wells inside of each tipbox. 
6) Seeds (which had been soaked for 2 hours in 1% bleach, rinsed, and then soaked overnight in pure H20) were placed in each well of each tipbox. One seed per well.
7) Watering occurred everyday to ensure that each container had an equal amount of water.
8) After two weeks, measurements were taken and roots were collected.

Below you can see an example container with rice seedlings growing. 
![vigor expeirment](Figures/vigor_pic.jpeg)

In a sense, there were two experiments going on here: a biological experiment to test whether microbiomes can affect seedling vigor, and a technical experiment to see if we could recapitulate different microbiomes across a common soil substrate.

## Data analysis

This will begin by analyzing the seedling vigor traits between the two soil types (wild vs. domesticated) when the inoculum is live or sterilized.

### Load in the necessary libraries
```{r}
library(tidyverse)
library(broom)
library(nlme)
library(multcomp)
```

### Load in the vigor data and format it
```{r}
seedling_data <- read_tsv("~/RMB/SoilDomestication/Data/seedling_vigor.tsv")

seedling_data <- seedling_data %>% 
  mutate(Trt = gsub("\\(-\\)", "Autoclaved Inoculum", Trt)) %>% 
  mutate(Trt = gsub("-", "Autoclaved Inoculum", Trt)) %>% 
  mutate(Trt = gsub("\\(\\+\\)", "Live Inoculum", Trt)) %>% 
  mutate(Type = ifelse(endsWith(Soil, "A"), "Domesticated", "Wild")) %>% 
  mutate(BinNumber = factor(BinNumber)) %>% 
  mutate(interactions = interaction(Trt, Type))
```

```{r}
seedling_data %>% 
  ggplot(aes(Type, Height, color = Trt, shape = Soil)) +
  geom_point(position = position_jitterdodge(jitter.width = 0.1)) +
  scale_color_manual(values = c("grey50", "darkseagreen")) +
  theme_minimal() +
  labs(x = "", y = "Height (cm)", color = "Treatment") +
  theme(text = element_text(size = 12))
```
With the domesticated soils, it appears that the seedlings growing in the live culture inoculated soils are shorter than the plants with the wild soil microbiome. Also with the autoclaved inoculum, there appears to be similar distribution in plant heights between the soils. We will look to see whether these effects are signficant below.

We next look at plant mass.
```{r}
seedling_data %>% 
  ggplot(aes(Type, Mass, color = Trt, shape = Soil)) +
  geom_point(position = position_jitterdodge(jitter.width = 0.1)) +
  scale_color_manual(values = c("grey50", "darkseagreen")) +
  theme_minimal() +
  labs(x = "", y = "Mass (g)", color = "Treatment") +
  theme(text = element_text(size = 12))
```
Similar story.

Let's model these results. In this particular experiment, it is important to keep in mind that several of the factors were nested inside other factors and the replication (two boxes per sample type) could add random noise to the results. Therefore, we will use a random effects model to test whether the soil type (wild vs. domesticated) affects seedling height and mass in the live culture and autoclaved microbiomes.
```{r}
height_model <- lme(Height ~ interactions, random = ~1|Type/Soil/BoxNumber, data =  seedling_data)
height_posthoc <- glht(height_model, linfct = mcp(interactions = "Tukey"))
tidy(summary(height_posthoc)) %>% mutate(p_adj = p.adjust(p.value, "bon"))
```
From the post-hoc comparisons, we see two important things regarding seedling height: plants growing in soils with live inocula have signficantly different heights between wild and domesticated soils. Plants grown in wild soil are on average 6.5 cm **taller** than plants grown in domesticated soils. This goes against the initial hypothesis stated in the introduction. Secondly, there are no signficant differences in heights between plants grown in soil with autoclaved inocula. Therefore, we can infer that the differences whitnessed between wild and domesticated soils with live inocula are due to biotic rather than edaphic differences between the soils.

We will now repeat this model using seedling mass as the dependent variable.

```{r}
mass_model <- lme(Mass ~ interactions, random = ~1|Type/Soil/BoxNumber, data =  seedling_data)
mass_posthoc <- glht(mass_model, linfct = mcp(interactions = "Tukey"))
tidy(summary(mass_posthoc)) %>% mutate(p_adj = p.adjust(p.value, "bon"))
```

Again, we see that when plants given the live inocula separate have signficantly different masses with plants coming from wild soil having a mass of around 70 mgs greater than plants in domesticated soils.

## Microbiome Analysis
Great! We appear to have a biological result. This was not what I expected at first. I thought plants with a domesticated microbiome would have great seedling vigor traits. Instead, we observed the opposite. Plants growing in domesticated soil microbiomes had lower seedling vigor values than plants with wild soil microbiomes. There are also reasons that this could be true -  a build up of pathogenic microbes seems the most likely. It could also be a tradeoff. Maybe plants with a domesticated microbiome have less growth overall, but the microbiome gives them some benefit (which is yet to be observed) that the wild microbiota cannot.

Anyway, none of the biological results mean anything if we cannot show that the microbiota established itself in a similar manner to plants grown in actual soil - rather than inoculated potting soil. We need to do a microbiome analysis to show this.

Here we go.

### Load in the libraries
```{r}
library(vegan)
library(reshape2)
library(tidyMB)
```

### Load in the relevant data
```{r}
tax <- read_rds("~/RMB/Reference/gg_otus_tax.rds")
vData <- wExp %>% 
  filter(study == "Seedling Vigor 1" | study == "Seedling Vigor 2") %>% 
  mutate(Compartment = gsub("ES", "Endosphere", Compartment), Compartment = gsub("RS", "Rhizosphere", Compartment)) %>% 
  mutate(Soil = gsub("AA", "ArbA", Soil), Soil = gsub("AN", "ArbN", Soil)) %>% 
  mutate(Soil = gsub("BA", "BiggsA", Soil), Soil = gsub("BN", "BiggsN", Soil)) %>% 
  mutate(Soil = gsub("D", "Davis", Soil), Soil = gsub("S", "SacN", Soil))
write_rds(vData, "~/RMB/SoilDomestication/Data/vigor_data.rds")
```

### Checking sequencing depth for each library
```{r}
vData %>% 
  group_by(SampleID, Compartment, study) %>% 
  summarise(depth = sum(value)) %>% 
  ggplot(aes(depth, fill = Compartment)) +
  geom_histogram() +
  facet_grid(study~Compartment) +
  scale_x_log10()
```

Looks pretty good. There is nothing too low.

### More cleaning, formatting, and saving
Remove OTUs that do not show up in at least 5% of the samples
```{r}
vData <- vData %>% 
  group_by(SampleID) %>% 
  mutate(Depth = sum(value), RA = (value / Depth) * 1000) %>%
  group_by(variable) %>% 
  filter(sum(value > 0) / n() > 0.05) %>% 
  select_if(~!all(is.na(.)))

write_rds(vData, "~/RMB/SoilDomestication/Data/vigor_data.rds")
vData <- read_rds("~/RMB/SoilDomestication/Data/vigor_data.rds")

vData <- vData %>% 
  mutate(Type = ifelse(grepl("A$", Soil), "Domesticated", "Wild")) %>% 
  mutate(Status = gsub("\\+", "Live Inoculum", Status)) %>% 
  mutate(Status = gsub("-", "Autoclaved Inoculum", Status))

```


### PCoAs
Every analysis needs a good PCoA to figure out what is going on.
```{r}
v_pc <- vData %>% 
  #filter(Soil != "BiggsA") %>% 
  filter(study == "Seedling Vigor 2") %>% 
  mutate(logRA = log2(RA + 1)) %>% 
  group_by(variable) %>% nest() %>% 
  mutate(variance = map_dbl(data, ~var(.x$logRA))) %>% 
  arrange(-variance) %>% 
  head(1000) %>% 
  unnest() %>% 
  group_by(Compartment, SampleType) %>% nest() %>% 
  mutate(pc = map(data, ~tidy_pcoa(.,value = "logRA")$axes)) %>% 
  unnest(pc)

v_pc %>% 
  ggplot(aes(MDS1, MDS2, color = Soil, shape = Type)) +
  geom_point(size = 3) +
  scale_color_manual(values = c("#92324B","#9e66a7",'#ee5768', '#fdc248', '#1ca04f','#27afb9')) +
  theme_minimal() +
  facet_wrap(SampleType~Compartment, scale = "free")

v_cap <- vData %>% 
  filter(SampleType == "Ino") %>% 
  filter(study == "Seedling Vigor 2" & SampleType != "Inoc") %>% 
  mutate(logRA = log2(RA + 1)) %>% 
  group_by(1) %>% nest() %>% 
  mutate(pc = map(data, ~tidy_cap(.,value = "logRA", formula = "1")$axes)) %>% 
  unnest(pc)

v_cap %>% 
  filter(SampleType != "PC") %>% 
  ggplot(aes(MDS1, MDS2, color = Soil, shape = SampleType)) +
  geom_point(size = 3) +
  scale_color_manual(values = c("#92324B","#9e66a7",'#ee5768', '#fdc248', '#1ca04f','#27afb9')) +
  theme(panel.background = element_rect(fill = "white"),
        panel.border = element_rect(fill = NA),
        axis.ticks.x = element_blank(),
        axis.ticks.y = element_blank())


```
The top PCoA shows the differences between the compartments and treatment (live vs dead inoculum). The "Pre" category is the micrbiome of the actual inoculum that went into the soils.

The bottom PCoA shows the differences between soil history. Notice that at the bottom of the graph are the autoclaved inocula samples. Their microbiomes of wild and domesticated soils in the autoclaved treatment appears to be converged.


```{r}
vData %>% 
  group_by(SampleID, Compartment, Status, Type) %>% 
  summarise(richness = sum(value > 0)) %>% 
  ungroup() %>% 
  mutate(Compartment = fct_relevel(Compartment, "Soil", "Rhizosphere", "Endosphere")) %>% 
  ggplot(aes(Compartment, richness, color = Type)) +
  geom_point(position = position_dodge(width = 0.5)) +
  facet_grid(.~Status, scale = "free_x") +
  scale_color_manual(values = c("dodgerblue", "gold")) +
  theme_minimal() +
  labs(y = "Species Richness", x = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1))
```


```{r}
vData %>% 
  mutate(logRA = log2(RA + 1)) %>% 
  #filter(Status != "Pre") %>% 
  group_by(Status, Compartment) %>% 
  nest() %>% 
  mutate(ad = map(data, ~long_adonis(., value = "logRA", formula = "Type"))) %>% 
  unnest(map(ad, ~tidy(.))) %>% 
  filter(term == "Type") %>% 
  mutate(Compartment = fct_relevel(Compartment, "Soil", "Rhizosphere", "Endosphere")) %>% 
  ggplot(aes(paste(Status, Compartment), R2, fill = Status)) +
  geom_bar(stat = "identity") +
  labs(y = "Soil history R2", x = "") +
  facet_grid(.~Compartment, scales = "free_x", space = "free_x") +
  scale_fill_manual(values = c("grey50", "darkseagreen", "mediumpurple1")) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 30, hjust = 1))
```
This result is showing the R^2^ values for soil history (wild vs domesticated) in each compartment and between autoclaved vs live inoculum. We can see that the strength of soil history on the acquired communities is much stronger with the live inoculum compared to the autoclaved. 

```{r}
vData %>% 
  inner_join(tax, by = "variable") %>% 
  group_by(Phylum2, SampleID, Compartment, Status, Soil, Type) %>% 
  summarise(total = sum(RA)) %>% 
  group_by(Phylum2) %>% nest() %>% 
  mutate(totz = map_dbl(data, ~sum(.x$total))) %>% 
  top_n(10, totz) %>% 
  dplyr::select(-totz) %>% 
  unnest() %>% 
  group_by(Compartment, SampleID, Status, Soil, Type) %>% 
  nest() %>% 
  group_by(Compartment, Soil, Type, Status) %>% 
  mutate(order = 1:n()) %>% 
  unnest() %>% 
  group_by(Compartment, Soil, Type, Status, Phylum2) %>% 
  summarise(mean_ab = mean(total)) %>% 
  ggplot(aes(paste(Type, Soil), mean_ab, fill = Phylum2)) +
  geom_bar(stat = "identity") +
  facet_grid(Compartment ~ Status, scales = "free_x") +
  scale_fill_brewer(palette = "Spectral") +
  labs(x = "", y = "Mean Proportion") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1))
```


### Differential abundance tests

In our previous results, we showed that the microbes enriched from the domesticated soils overlap to a considerable with rice enriched microbes. We will again test this, but this time using the inoculated soils. Ideally, we would expect that plants growing in the soils with the live inocula should also host rice specific microbes while plants growing the soils with the autoclaved inocula should not.

The approach we will take here will be to model each compartment and treatment (live vs autoclaved) separately to identify OTUs differentially abundant between soil history types. Again, there are two "wild" soils and two "domesticated" soils. We will compare each wild to each domesticated soil and find the OTUs that are differentially abundand across each comparison.
```{r}
library(stringr)
library(DESeq2)
library(biobroom)
vDE <- vData %>% 
  group_by(variable) %>% 
  filter(sum(value > 0) / n() >= 0.1) %>% 
  filter(Status != "Pre") %>% 
  mutate(group = factor(paste(Compartment, Soil, sep = "."))) %>% 
  group_by(Status) %>% 
  nest() %>% 
  mutate(DGEL = map(data, ~suppressMessages(tidyDGEL(., method = "DESeq2", value = "value", group_column = "group", formula = "~ group")))) %>% 
  mutate(dds = map(DGEL, ~suppressMessages(DESeq(.)))) %>% 
  mutate(BulkSoil_ArbA_ArbN = map(dds, ~lfcShrink(., contrast = c("group", "Soil.ArbA", "Soil.ArbN")))) %>% 
  mutate(BulkSoil_ArbA_SacN = map(dds, ~lfcShrink(., contrast = c("group", "Soil.ArbA", "Soil.SacN")))) %>% 
  mutate(BulkSoil_BiggsA_ArbN = map(dds, ~lfcShrink(., contrast = c("group", "Soil.BiggsA", "Soil.ArbN")))) %>% 
  mutate(BulkSoil_BiggsA_SacN = map(dds, ~lfcShrink(., contrast = c("group", "Soil.BiggsA", "Soil.SacN")))) %>% 
  mutate(Rhizosphere_ArbA_ArbN = map(dds, ~lfcShrink(., contrast = c("group", "Rhizosphere.ArbA", "Rhizosphere.ArbN")))) %>% 
  mutate(Rhizosphere_ArbA_SacN = map(dds, ~lfcShrink(., contrast = c("group", "Rhizosphere.ArbA", "Rhizosphere.SacN")))) %>% 
  mutate(Rhizosphere_BiggsA_ArbN = map(dds, ~lfcShrink(., contrast = c("group", "Rhizosphere.BiggsA", "Rhizosphere.ArbN")))) %>% 
  mutate(Rhizosphere_BiggsA_SacN = map(dds, ~lfcShrink(., contrast = c("group", "Rhizosphere.BiggsA", "Rhizosphere.SacN")))) %>%
  mutate(Endosphere_ArbA_ArbN = map(dds, ~lfcShrink(., contrast = c("group", "Endosphere.ArbA", "Endosphere.ArbN")))) %>% 
  mutate(Endosphere_ArbA_SacN = map(dds, ~lfcShrink(., contrast = c("group", "Endosphere.ArbA", "Endosphere.SacN")))) %>% 
  mutate(Endosphere_BiggsA_ArbN = map(dds, ~lfcShrink(., contrast = c("group", "Endosphere.BiggsA", "Endosphere.ArbN")))) %>% 
  mutate(Endosphere_BiggsA_SacN = map(dds, ~lfcShrink(., contrast = c("group", "Endosphere.BiggsA", "Endosphere.SacN")))) %>% 
  dplyr::select(BulkSoil_ArbA_ArbN, BulkSoil_ArbA_SacN, BulkSoil_BiggsA_ArbN, BulkSoil_BiggsA_SacN,
                Rhizosphere_ArbA_ArbN, Rhizosphere_ArbA_SacN, Rhizosphere_BiggsA_ArbN, Rhizosphere_BiggsA_SacN,
                Endosphere_ArbA_ArbN, Endosphere_ArbA_SacN, Endosphere_BiggsA_ArbN, Endosphere_BiggsA_SacN, Status) %>% 
  gather(key = Comparison, value = results, -Status)

write_rds(vDE, "~/RMB/SoilDomestication/Data/vigorDA.rds")
```

How many OTUs are differentially abundant between soil history across the soil treatments (sterilized vs not sterilized)?
```{r}
vDE <- read_rds("~/RMB/SoilDomestication/Data/vigorDA.rds")
vDE %>% 
  unnest(map(results, ~tidy(.))) %>% 
  separate(Comparison, into = c("Compartment", "Soil1", "Soil2"), sep = "_") %>% 
  mutate(direction = ifelse(estimate > 0, "A", "N")) %>% 
  filter(p.adjusted <= 0.05) %>% ungroup() %>% 
  dplyr::count(Compartment, Status, gene, direction) %>% 
  filter(n == 4) %>% 
  dplyr::count(Compartment, Status, direction) %>% 
  mutate(group = factor(paste(Compartment, Status))) %>% 
  #mutate(group = fct_relevel(group, "BulkSoil -", "BulkSoil +", "Rhizosphere -", "Rhizosphere +",
  #                           "Endosphere -", "Endosphere +")) %>% 
  ggplot(aes(group, nn, fill = Status)) +
  geom_bar(stat = "identity", width = 0.9) +
  scale_fill_manual(values = c("grey50", "darkseagreen")) +
  theme_minimal() +
  labs(x = "", y = "Differentially Abundant OTUs Between Wild and Domesticated Soils") +
  theme(axis.text.x = element_text(angle =  30, hjust = 1), text = element_text(size = 12),
        legend.position = "none") 
```
This graph shows that there are many OTUs differentially abundant in the live inoculum soils but very few in the autoclaved soils. This is exactly what we would expect. Because the this experiment was not performed in a closed system, we would expect that the microbes the roots are aquiring microbes from the environment, but the assembled communities are either going to be consistent or random and therefore we would not find differentially abundant OTUs between soil histories. We know from the PCoA that these communities residing in the soil with autoclaved inocula are consistent.

Let us next take a peek at the which phyla the differentially abundant OTUs belong to. 
```{r}
vDE %>% 
  unnest(map(results, ~tidy(.))) %>% 
  separate(Comparison, into = c("Compartment", "Soil1", "Soil2"), sep = "_") %>% 
  mutate(direction = ifelse(estimate > 0, "Domesticated", "Wild")) %>% 
  filter(p.adjusted <= 0.05) %>% ungroup() %>% 
  dplyr::count(Compartment, Status, gene, direction) %>% 
  filter(n == 4) %>% 
  inner_join(tax, by = c("gene" = "variable")) %>% 
  dplyr::count(Compartment, Status, direction, Phylum2) %>% 
  mutate(Compartment = fct_relevel(Compartment, "BulkSoil", "Rhizosphere", "Endosphere")) %>% 
  ggplot(aes(Compartment, y = nn, fill = Phylum2)) +
  geom_bar(stat = "identity") +
  facet_grid(.~Status + direction, scales = "free_x") +
  scale_fill_manual(values = c(brewer.pal(11, "Spectral"), 
                               brewer.pal(4, "Reds"),
                               brewer.pal(4, "Blues"))) +
  labs(x = "", y = "Number of OTUs") +
  theme_minimal() +
  guides(fill = guide_legend(ncol = 2)) +
  theme(axis.text.x = element_text(angle = 30, hjust = 1))
```
This is promising. If we go back and look at SFig9, we can see that the domesticated OTUs belong to a few phyla - primarily Deltaproteobacteria, Acidobacteria, and Chloroflexi, among others. In the wild soils we see that they have OTUs which belong to Gammaproteobacteria, and some others. The graph above appears to be showing the same trends.

Let's next look at the overlap of rice-enriched OTUs compared to the OTUs we found to be enriched in the domesticated soils in this experiment.

```{r}
np_deseq <- read_rds("~/RMB/SoilDomestication/Data/weeds_deseq.rds")
rice_otus <- np_deseq %>% 
  unnest(map(results, ~tidy(.))) %>% 
  separate(taxon, sep = "_", into = c("Taxon", "Compartment")) %>% 
  mutate(direction = ifelse(estimate < 0, "Weed", "Rice")) %>% 
  filter(p.adjusted <= 0.05) %>% 
  dplyr::count(direction, Compartment, gene) %>% 
  dplyr::add_count(direction, Compartment) %>% 
  filter(n == 3) %>% 
  mutate(Compartment = gsub("ES", "Endosphere", Compartment),
         Compartment = gsub("RS", "Rhizosphere", Compartment)) %>% 
  dplyr::rename(variable = gene) %>% 
  filter(direction == "Rice") %>% 
  dplyr::count(variable)

dom_deseq <- readRDS("~/RMB/SoilDomestication/Data/domestication_deseq.rds")
dom_otus <- dom_deseq %>% 
  unnest(map(results, ~tidy(.))) %>% 
  group_by(Experiment, Compartment) %>% 
  mutate(direction = ifelse(estimate > 0, "A", "N")) %>% 
  filter(p.adjusted <= 0.05) %>% ungroup() %>% 
  dplyr::count(direction, Compartment, gene) %>% 
  filter(n == 2) %>% 
  dplyr::count(direction, gene, Compartment) %>% 
  dplyr::rename(variable = gene) %>% 
  filter(direction == "A") %>% 
  dplyr::count(variable)
  
A_otus <- vDE %>% 
  unnest(map(results, ~tidy(.))) %>% 
  separate(Comparison, into = c("Compartment", "Soil1", "Soil2"), sep = "_") %>% 
  mutate(direction = ifelse(estimate > 0, "Domesticated", "Wild")) %>% 
  filter(p.adjusted <= 0.05) %>% ungroup() %>% 
  dplyr::count(Compartment, Status, gene, direction) %>% 
  filter(n == 4 & direction == "Domesticated" & Status == "Live Inoculum") %>% 
  dplyr::count(gene)
```

```{r}
library(phyloseq)
gg_tree <- readRDS("~/RMB/SoilDomestication/Data/gg_tree.rds")

in_otus <- vDE %>% 
  unnest(map(results, ~tidy(.))) %>% 
  separate(Comparison, into = c("Compartment", "Soil1", "Soil2"), sep = "_") %>% 
  mutate(direction = ifelse(estimate > 0, "Domesticated", "Wild")) %>% 
  filter(p.adjusted <= 0.05) %>% ungroup() %>% 
  dplyr::count(Compartment, Status, gene, direction) %>% 
  filter(n == 4 & Status == "Live Inoculum") %>% 
  mutate(variable = gene)

domMD <- data.frame(Compartment = c("aShared", rep(c("bBulkSoil", "cRhizosphere", "dEndosphere"), each = 2)),
                      Type = c("aShared",rep(c("Domesticated", "Wild"), 3)),
                      SampleID = paste("D", 0:6, sep = ""),
                    row.names = paste("D", 0:6, sep = ""))

dT <- in_otus %>% 
  mutate(Type = direction) %>% 
  mutate(Compartment = gsub("BulkSoil", "bBulkSoil", Compartment),
         Compartment = gsub("Rhizosphere", "cRhizosphere", Compartment),
         Compartment = gsub("Endosphere", "dEndosphere", Compartment)) %>% 
  bind_rows(rice_otus %>% filter(variable %in% 
                                   intersect(in_otus %>% 
                                               filter(direction == "Domesticated") %>% 
                                               pull(variable), rice_otus$variable)) %>% 
              dplyr::count(variable) %>% 
              mutate(Type = "aShared", Compartment = "aShared")) %>% 
  mutate(value = 1) %>% 
  inner_join(domMD, by = c("Compartment", "Type")) %>%
  dplyr::select(SampleID, variable, value) %>% 
  spread(SampleID, value, fill = 0) %>% 
  as.data.frame()
row.names(dT) <- dT$variable
dT$variable <- NULL  

dTax <- in_otus %>% 
  inner_join(tax, by = "variable") %>% 
  spread(Compartment, direction) %>% 
  mutate(Phylum2 = as.character(Phylum2)) %>% 
  dplyr::rename(Domain = Kingdom) %>% 
  as.data.frame()
row.names(dTax) <- dTax$variable
dTax <- dTax[match(row.names(dT), row.names(dTax)),]


dom.physeq_sampleType <- phyloseq(otu_table(dT, taxa_are_rows = T), 
                       sample_data(domMD),
                       dTax,
                       gg_tree)
# Plot tree with metadata
plot_tree(dom.physeq_sampleType, color = "Type", ladderize = "left", shape = "Compartment", justify = "left", size = "Type") +
  scale_color_manual(values = c('black', "dodgerblue", "gold")) +
  scale_shape_manual(values = c(1, 16, 15, 17)) +
  scale_size_manual(values = c(1,2,2,2))

# Plot tree with taxonomic information
dom.physeq_Phy <- phyloseq(otu_table(data.frame(J1 = rep(1, nrow(dT)), row.names = row.names(dT)), taxa_are_rows = T), 
                       tax_table(as.matrix(dTax)),
                       gg_tree)

plot_tree(dom.physeq_Phy, color = "Phylum2", ladderize = "left") 
```


```{r}
A_overlap <- intersect(rice_otus$variable, A_otus$gene)
dom_overlap <- intersect(rice_otus$variable, dom_otus$variable)
dov_v_vig_overlap <- intersect(dom_otus$variable, A_otus$gene)

print(paste(length(A_overlap), "OTUs overlap between dometicated OTUs (this expeirment) and the rice enriched microbes."))
print(paste(length(dom_overlap), "OTUs overlap between dometicated OTUs (previous expeirment) and the rice enriched microbes."))
print(paste(length(dov_v_vig_overlap), "OTUs overlap between dometicated OTUs in this expeirment and the domesticated OTUs from the previous experiment."))
```

This is very exciting. Of the OTUs found to be enriched in plants growing in the domesticated soils, 42 of them overlap with the rice enriched microbes from the Arkansas experiment. This is on par with what we discovered in the previous experiment.



