---
title: "Native Plants Analysis"
output: html_notebook
author: "Joe Edwards"
---

```{r}
library(vegan)
library(MASS)
library(tidyverse)
library(broom)
library(biobroom)
library(tidyMB)
library(RColorBrewer)
library(DESeq2)
```

## Load in a tidy version of the count data and other necessary dataframes
```{r}
nExp <- readRDS("~/RMB/SoilDomestication/Data/native_plants_data.rds") %>% 
  mutate(cpm = round(value * (1000000 / depth)))
tax <- readRDS("~/RMB/SoilDomestication/Data/gg_otus_tax.rds")
```

## Get variance stabilized values for each OTU in each sample for downstream analysis
```{r}
nmap <- nExp %>% 
  dplyr::select(variable, Compartment, SampleID, host_common_name, Site, Experiment, value) %>% 
  grab_metadata(., samples = "SampleID", otus = "variable")

nExp_vsd <- nExp %>% 
  dplyr::select(variable, SampleID, value, host_common_name, study, Site, Compartment, Experiment) %>% 
  group_by(variable) %>% filter(sum(value > 0) / n() > 0.05) %>% 
  group_by(1) %>% nest() %>% 
  mutate(dds = map(data, ~tidyDGEL(., method = "DESeq2", samples = "SampleID", formula = "~ host_common_name"))) %>% 
  mutate(vsd = map(dds, ~assay(varianceStabilizingTransformation(., blind = T))))

vsd_table <- nExp_vsd$vsd[[1]]

nExp_vsd <- inner_join(nExp_vsd$vsd[[1]] %>% 
                         as.data.frame() %>% 
                         rownames_to_column("variable") %>% 
                         gather("SampleID", "vsd", -variable), nmap, by = "SampleID")
```

## Perform weighted and unweighted unifrac analyses
```{r}
library(ape)
library(GUniFrac)

# convert negative values to zeros
vsd_table[vsd_table < 0] <- 0
vsd_table <- t(vsd_table)

# Load in tree
tree <- read.tree("~/RMB/Reference/97_otus.tree")

## Subset tree to only have otus from vsd table
good_tree <- keep.tip(tree, colnames(vsd_table))

## One edge length is NaN, throwing off the calculations below. So impute it by taking the average
good_tree$edge.length[is.na(good_tree$edge.length)] <- mean(good_tree$edge.length, na.rm = T)

## Calculate unifrac
unifracs <- GUniFrac(vsd_table, good_tree, alpha=c(0, 0.5, 1))$unifracs

## Calculate pcoa for each dist metric
wuf_pc <- capscale(unifracs[,,"d_1"] ~ 1)
uuf_pc <- capscale(unifracs[,,"d_UW"] ~ 1)

## Plot pcoas
uuf_plot <- uuf_pc$CA$u %>% 
  as.data.frame() %>% 
  rownames_to_column("SampleID") %>% 
  inner_join(nmap) %>% 
  ggplot(aes(MDS1, MDS2, color = host_common_name, shape = Compartment)) +
  geom_point(size = 3, alpha = 0.9) +
  #stat_ellipse(aes(group = Compartment), color = "black") +
  scale_color_brewer(palette = "Set1", direction = -1) +
  scale_shape_manual(values = c(1, 18, 16)) +
  labs(x = paste("PCo1 (", round(uuf_pc$CA$eig[1]/sum(uuf_pc$CA$eig) * 100, 2), "%)", sep = ""), 
       y = paste("PCo2 (", round(uuf_pc$CA$eig[2]/sum(uuf_pc$CA$eig) * 100, 2), "%)", sep = "")) +
  theme(text = element_text(size = 12),
        panel.background = element_rect(fill = "white"),
        panel.border = element_rect(fill = NA),
        axis.ticks.x = element_blank(),
        axis.ticks.y = element_blank(),
        legend.position = "none")

wuf_plot <- wuf_pc$CA$u %>% 
  as.data.frame() %>% 
  rownames_to_column("SampleID") %>% 
  inner_join(nmap) %>% 
  ggplot(aes(MDS1, MDS2, color = host_common_name, shape = Compartment)) +
  geom_point(size = 3, alpha = 0.9) +
  #stat_ellipse(aes(group = Compartment), color = "black") +
  scale_color_brewer(palette = "Set1", direction = -1) +
  scale_shape_manual(values = c(1, 18, 16)) +
  labs(x = paste("PCo1 (", round(wuf_pc$CA$eig[1]/sum(wuf_pc$CA$eig) * 100, 2), "%)", sep = ""), 
       y = paste("PCo2 (", round(wuf_pc$CA$eig[2]/sum(wuf_pc$CA$eig) * 100, 2), "%)", sep = "")) +
  theme(text = element_text(size = 12),
        panel.background = element_rect(fill = "white"),
        panel.border = element_rect(fill = NA),
        axis.ticks.x = element_blank(),
        axis.ticks.y = element_blank(),
        legend.position = "none")

cowplot::plot_grid(wuf_plot, uuf_plot)
```

## Do permanova on the unifrac distance matrices
```{r}
adonis(as.dist(unifracs[,,"d_UW"]) ~ Compartment * host_common_name, data = nmap)$aov.tab %>% tidy() %>% View()
adonis(as.dist(unifracs[,,"d_1"]) ~ Compartment * host_common_name, data = nmap)$aov.tab %>% tidy() %>% View()
```

## Perform PCoA on variance stabilized data using Bray-Curtis dissimilarities
```{r} 
nPC <- tidy_pcoa(nExp_vsd %>% mutate(vsd = ifelse(vsd < 0, 0, vsd)), dist = "bray", value = "vsd", keep_loadings = T)

nPC_plot <- nPC$axes %>% 
  ggplot(aes(MDS1, MDS2, color = host_common_name, shape = Compartment)) +
  geom_point(size = 3, alpha = 0.9) +
  #stat_ellipse(aes(group = Compartment), color = "black") +
  scale_color_brewer(palette = "Set1", direction = -1) +
  scale_shape_manual(values = c(1, 18, 16)) +
  labs(x = paste("PCo1 (", round(nPC$eigen_vals[1] * 100, 2), "%)", sep = ""), 
       y = paste("PCo2 (", round(nPC$eigen_vals[2] * 100, 2), "%)", sep = "")) +
  theme(text = element_text(size = 12),
        panel.background = element_rect(fill = "white"),
        panel.border = element_rect(fill = NA),
        axis.ticks.x = element_blank(),
        axis.ticks.y = element_blank())

nPC_plot
```

## Perform ANOVA on each axis to detect contribution of compartment and host species
```{r}
nPC$axes %>% 
  dplyr::select(SampleID, host_common_name, Compartment, MDS1:MDS5) %>% 
  gather(key = Axis, value = value, -c(SampleID, host_common_name, Compartment)) %>% 
  group_by(Axis) %>% filter(Compartment != "Bulk Soil") %>% 
  nest() %>% 
  mutate(mod = map(data, ~aov(value ~ Compartment + host_common_name, .))) %>% 
  unnest(map(mod, ~tidy(.))) %>% 
  filter(term == "host_common_name")
```

## Do as above, but this time remove rice since it behaves differently than the other host plants
```{r}
##
nPC$axes %>% 
  dplyr::select(SampleID, host_common_name, Compartment, MDS1:MDS5) %>% 
  gather(key = Axis, value = value, -c(SampleID, host_common_name, Compartment)) %>% 
  group_by(Axis) %>% filter(Compartment != "Bulk Soil" & host_common_name != "Rice") %>% 
  nest() %>% 
  mutate(mod = map(data, ~aov(value ~ Compartment + host_common_name, .))) %>% 
  unnest(map(mod, ~tidy(.))) %>% 
  filter(term == "host_common_name")
```


## Compare dissimilarities to bulk soil samples
```{r}
np_dist <- long_distance(nExp_vsd %>% mutate(vsd = ifelse(vsd < 0, 0, vsd)), value = "vsd")
np_dist %>% 
  filter(Compartment.x == "Bulk Soil" | Compartment.y == "Bulk Soil") %>% 
  mutate(name = gsub("Soil", "", paste0(host_common_name.x, host_common_name.y))) %>% 
  mutate(Compartment = gsub("Bulk Soil", "", paste0(Compartment.x, Compartment.y))) %>% 
  mutate(name = ifelse(name == "", "Soil", name)) %>%
  mutate(Compartment = ifelse(Compartment == "", "Rhizosphere", Compartment)) %>% 
  mutate(Compartment = fct_relevel(Compartment, "Bulk Soil", "Rhizosphere", "Endosphere")) %>% 
  mutate(name = fct_relevel(name, "Soil", "Rice", "Sedge", "Mudplantain", "Redstem")) %>% 
  filter(Compartment == "Rhizosphere" & value <= 1) %>% 
  ggplot(aes(name, value)) +
  geom_boxplot(outlier.size = 0.1) +
  geom_jitter(width = 0.2, alpha = 0.2, size = 0.5) +
  labs(y = "Bray-Curtis dissimilarity to bulk soil") +
  #facet_grid(.~Compartment) +
  theme_minimal() +
  theme(text = element_text(size = 12))

np_dist %>% 
  filter(Compartment.x != Compartment.y) %>% 
  filter(Compartment.x == "Bulk Soil" | Compartment.y == "Bulk Soil") %>% 
  mutate(name = gsub("Soil", "", paste0(host_common_name.x, host_common_name.y))) %>% 
  mutate(Compartment = gsub("Bulk Soil", "", paste0(Compartment.x, Compartment.y))) %>% 
  mutate(Compartment = fct_relevel(Compartment, "Rhizosphere", "Endosphere")) %>% 
  mutate(name = fct_relevel(name, "Rice", "Sedge", "Mudplantain", "Redstem")) %>% 
  filter(Compartment == "Rhizosphere") %>% 
  group_by(Compartment) %>% 
  nest() %>% 
  mutate(mod = map(data, ~aov(value ~ name, .))) %>% 
  unnest(map(mod, ~tidy(TukeyHSD(.)))) %>% View()
```


```{r}
np_dist %>% 
  filter(Compartment.x != "Bulk Soil" & Compartment.y != "Bulk Soil") %>% 
  filter(Compartment.x == Compartment.y) %>% 
  filter(Compartment.x == "Endosphere") %>% 
  group_by(host_common_name.x, host_common_name.y, Compartment.x) %>% 
  summarise(mean_dis = mean(value)) %>% 
  mutate(plant1 = ifelse(host_common_name.x > host_common_name.y, host_common_name.x, host_common_name.y)) %>% 
  mutate(plant2 = ifelse(host_common_name.x < host_common_name.y, host_common_name.x, host_common_name.y)) %>%
  ungroup() %>% 
  mutate(plant1 = fct_relevel(plant1, "Mudplantain", "Redstem", "Sedge", "Rice")) %>% 
  mutate(plant2 = fct_relevel(plant2, "Mudplantain", "Redstem", "Sedge", "Rice")) %>% 
  mutate(plant1)
  ggplot(aes(plant1, plant2, fill = mean_dis)) +
  geom_tile(color = "black") +
  facet_grid(.~Compartment.x) +
  scale_fill_gradient(low = "white", high = "red") +
  theme_minimal()
```

## Perform permanova with the Bray-Curtis dissimilarities
```{r}
long_adonis(nExp_vsd, value = "vsd", formula = "Compartment * host_common_name") %>% tidy()
```
```{r}
nExp_vsd %>% 
  group_by(Compartment) %>% 
  filter(Compartment != "Bulk Soil") %>% nest() %>% 
  mutate(adon = map(data, ~long_adonis(., value = "vsd", formula = "host_common_name"))) %>% 
  unnest(map(adon, ~tidy(.)))
```
```{r}
a_div <- nExp %>% 
  mutate(Compartment = fct_relevel(Compartment, "Bulk Soil", "Rhizosphere", "Endopshere")) %>% 
  mutate(host_common_name = relevel(factor(host_common_name), ref = "Rice")) %>% 
  group_by(Compartment, host_common_name, SampleID) %>% 
  filter(RA > 0) %>% 
  summarise(richness = sum(RA > 0), shannon = 2^(-sum(log2(RA) * RA))) 

a_div %>% 
  gather(key = metric, value = diversity, -c(Compartment, host_common_name, SampleID)) %>% 
  group_by(Compartment, metric) %>% 
  nest() %>% 
  filter(Compartment != "Bulk Soil") %>% 
  mutate(mod = map(data, ~aov(diversity ~ host_common_name, .))) %>% 
  unnest(map(mod, ~tidy(TukeyHSD(.))))

a_div %>% group_by(1) %>% nest() %>% 
  mutate(mod = map(data, ~aov(richness ~ Compartment * host_common_name, .))) %>% 
  unnest(map(mod, ~tidy(.)))

a_div %>% 
  ggplot(aes(Compartment, richness, color = host_common_name)) +
  geom_point(position = position_jitterdodge(jitter.width = 0.2, jitter.height = 0)) +
  scale_color_manual(values = c("#4DAF4A", "#FF7F00", "#984EA3", "#377EB8", "#E41A1C")) +
  theme_minimal() +
  labs(x = "", y = "Detectable OTUs") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1), legend.position = "none")
```

From the above PERMANOVA, host species does have a strong affect on the microbiome, but looking at the PCoA it is not clear whether the genotypic effect exists only because rice is so weird compared to the other plants.

```{r}
nExp_vsd %>% 
  filter(Compartment != "Bulk Soil" & host_common_name != "Rice") %>% 
  group_by(Compartment) %>% 
  nest() %>% 
  mutate(permanova = map(data, ~long_adonis(., value = "vsd", formula = "host_common_name"))) %>% 
  unnest(map(permanova, ~tidy(.)))
```

```{r}
nExp %>% 
  mutate(RA2 = log2((RA*1000) + 1)) %>% 
  group_by(Compartment) %>% 
  nest() %>% 
  filter(Compartment != "Bulk Soil") %>% 
  mutate(ad = map(data, ~long_adonis(., value = "RA2", formula = "host_common_name"))) %>% 
  unnest(map(ad, ~tidy(.)))
```
Here we have the permanova results for how host species affects the microbiome within each compartment. The top table shows the results when rice is excluded from the analysis. The lower model includes rice. By comparing the R2 values, we can see that excluding rice from the analysis drastically reduces the effect size, but the result is still significant.

## Phyla Stuff THIS HAS BEEN REMOVED FROM THE MANUSCRIPT
```{r}
phyla_abund <- nExp %>% 
  inner_join(tax, by = "variable") %>% 
  group_by(SampleID, Compartment, host_common_name, Phylum2, depth) %>% 
  summarise(phy_total = sum(value)) %>% 
  mutate(prop = (phy_total + 1) / (depth + 1))

phyla_abund %>% 
  group_by(Phylum2) %>% 
  nest() %>% 
  mutate(total = map_dbl(data, ~sum(.x$prop))) %>% 
  top_n(10, total) %>% 
  unnest(data) %>% 
  mutate(Compartment = fct_relevel(Compartment, "Bulk Soil", "Rhizosphere", "Endosphere")) %>% 
  group_by(Compartment, host_common_name, SampleID) %>% 
  nest() %>% 
  group_by(Compartment) %>% 
  arrange(host_common_name) %>% 
  mutate(order = 1:n()) %>% 
  unnest() %>% 
  ggplot(aes(order, prop * 100, fill = Phylum2)) +
  geom_bar(stat = "identity", width = 1) +
  geom_point(aes(x = order, y = -2, color = host_common_name), shape = 15) +
  facet_grid(.~Compartment, scales = "free_x") +
  #scale_fill_manual(values = c(brewer.pal(11, "RdGy"),brewer.pal(5, "Blues")[-1])) +
  scale_fill_manual(values = c(RColorBrewer::brewer.pal(6, "Spectral"), RColorBrewer::brewer.pal(6, "Greys")[-1])) +
  scale_color_brewer(palette = "Set1", direction = -1) +
  labs(x = "", y = "Percent of Reads") +
  theme_minimal() +
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        legend.position = "bottom",
        text = element_text(size = 12))

RColorBrewer::brewer.pal(12, "Spectral")
"#66C2A5" "#D53E4F" "#5E4FA2"
 "#D53E4F" "#F46D43" "#FDAE61" "#FEE08B" "#FFFFBF" "#E6F598" "#ABDDA4" "#66C2A5" "#3288BD" "#5E4FA2"
```

Let's use beta regression to find phyla whose relative abundance is signficantly different from rice. Betaregression requires that the dependent variable be in the [0,1] interval. If it's 0 or 1 it will not work. So the proportions that are being used here are calculated as (phylum counts within sample + 1) / (sequencing depth of sample + 1).
```{r}
library(betareg)
safe_betareg <- possibly(betareg, NA_real_)
nPhyModel_host <- phyla_abund %>% 
  ungroup() %>% 
  filter(Compartment != "Bulk Soil") %>% 
  mutate(host_common_name = relevel(factor(host_common_name), "Rice")) %>% 
  group_by(Phylum2, Compartment) %>% 
  nest() %>% 
  mutate(model = map(data, ~safe_betareg(prop ~ host_common_name, .))) %>% 
  unnest(map(model, ~tidy(.))) %>% 
  

nPhyModel_host %>% 
  filter(term != "(Intercept)" & term != "(phi)") %>% 
  group_by(Compartment) %>% 
  mutate(p.adj = p.adjust(p.value, "BH")) %>% 
  filter(p.adj <= 0.05) %>% 
  mutate(direction = ifelse(estimate > 0, "Weed", "Rice")) %>% 
  dplyr::count(direction, Compartment, Phylum2) %>% 
  filter(n == 3) %>% 
  dplyr::select(-n) %>% 
  write_tsv("~/RMB/SoilDomestication/Tables/enriched_phyla_native_plants.tsv")

nPhyModel_host %>% 
  filter(term != "(Intercept)" & term != "(phi)") %>% 
  group_by(Compartment) %>% 
  mutate(p.adj = p.adjust(p.value, "BH")) %>% 
  filter(p.adj <= 0.05) %>% 
  mutate(direction = ifelse(estimate > 0, "Other", "Rice")) %>% 
  dplyr::count(direction, Compartment, Phylum2) %>% 
  filter(n == 3) %>% 
  dplyr::count(direction, Compartment)
```


```{r}
nPhyModel_compartment <- phyla_abund %>% 
  ungroup() %>% 
  mutate(Compartment = fct_relevel(factor(Compartment), "Bulk Soil", "Rhizosphere", "Endosphere")) %>% 
  group_by(Phylum2) %>% 
  nest() %>% 
  mutate(model = map(data, ~safe_betareg(prop ~ Compartment, .))) %>% 
  unnest(map(model, ~tidy(.))) 

nPhyModel_compartment %>% 
  filter(term != "(Intercept)" & term != "(phi)") %>% 
  mutate(term = gsub("Compartment", "", term)) %>% 
  mutate(p.adj = p.adjust(p.value, "BH")) %>% 
  inner_join(phyla_abund %>% group_by(Phylum2, Compartment) %>% summarise(mean_ab = mean(prop)), by = c(c("term" = "Compartment"), "Phylum2")) %>% 
  mutate(Compartment = fct_relevel(term, "Rhizosphere", "Endosphere")) %>%
  ggplot(aes(Compartment, estimate, group = Phylum2, color = ifelse(p.adj <= 0.05, Compartment, "ns"), shape = ifelse(p.adj <= 0.05, "sig", "ns"), size = log2(mean_ab * 1000))) +
  geom_line(color = 'black', alpha = 0.5, size = 0.5) +
  geom_point(alpha = 0.5) +
  scale_shape_manual(values = c(1, 16)) +
  scale_color_manual(values = c("#DFC27D", "#80CDC1", "black")) +
  #scale_color_manual(values = c("darkmagenta", "steelblue", "black")) +
  scale_size_continuous(range = c(0,10), breaks = c(-4, 0, 4, 8)) +
  theme_minimal() +
  labs(x = "") +
  theme(text = element_text(size = 12))

```


## Differential OTU Abundance
Let's find the OTUs that are different between the microbiomes of the different plants. In this analysis I will limit the contrasts to comparisons between the other plants and rice.


```{r}
## I commented out this whole chunk because I don't want to run it again. I saved the results as an rds object and will read it in for working in the future.

DESeq2_plant <- nExp %>% 
   dplyr::select(SampleID, variable, value, Compartment, host_common_name) %>%
   filter(Compartment != "Bulk Soil") %>% 
   group_by(variable) %>% 
   filter(sum(value > 0) / n() > 0.1) %>% 
   mutate(host_common_name = relevel(factor(host_common_name), ref = "Rice")) %>% 
   mutate(group = factor(paste(Compartment, host_common_name, sep = "."))) %>% 
   group_by(group_var = "DESeq2") %>% 
   nest() %>% 
   mutate(DGEL = map(data, ~tidyDGEL(., value = "value", formula = "~group", method = "DESeq2"))) %>%
   mutate(dds = map(DGEL, ~DESeq(.))) %>% 
   mutate(Sedge_RS = map(dds, ~lfcShrink(., contrast = c("group", "Rhizosphere.Rice", "Rhizosphere.Sedge")))) %>% 
   mutate(Mudplantain_RS = map(dds, ~lfcShrink(., contrast = c("group", "Rhizosphere.Rice", "Rhizosphere.Mudplantain")))) %>% 
   mutate(Redstem_RS = map(dds, ~lfcShrink(., contrast = c("group", "Rhizosphere.Rice", "Rhizosphere.Redstem")))) %>% 
   mutate(Sedge_ES = map(dds, ~lfcShrink(., contrast = c("group", "Endosphere.Rice", "Endosphere.Sedge")))) %>% 
   mutate(Mudplantain_ES = map(dds, ~lfcShrink(., contrast = c("group", "Endosphere.Rice", "Endosphere.Mudplantain")))) %>% 
   mutate(Redstem_ES = map(dds, ~lfcShrink(., contrast = c("group", "Endosphere.Rice", "Endosphere.Redstem")))) %>% 
   dplyr::select(Mudplantain_RS, Sedge_RS, Redstem_RS, Mudplantain_ES, Sedge_ES, Redstem_ES, group_var) %>% 
   gather(key = taxon, value = results, -group_var)
```

```{r}
saveRDS(DESeq2_plant, file = "~/RMB/SoilDomestication/Data/native_plants_deseq.rds")
```

```{r}
DESeq2_plant <- read_rds("~/RMB/SoilDomestication/Data/native_plants_deseq.rds")

np_MA_plot <- DESeq2_plant %>% 
  unnest(map(results, ~tidy(.))) %>% 
  separate(taxon, sep = "_", into = c("Taxon", "Compartment")) %>% 
  mutate(direction = ifelse(estimate < 0, "Weed", "Rice")) %>% 
  mutate(p.adjusted = ifelse(is.na(p.adjusted), 1, p.adjusted)) %>% 
  mutate(sig = ifelse(p.adjusted <= 0.05, "sig", "ns")) %>% 
  mutate(sig_direction = ifelse(p.adjusted <= 0.05, direction, NA)) %>% 
  mutate(color = ifelse(sig_direction == "Rice", "Rice", Taxon)) %>% 
  mutate(color = ifelse(is.na(color), "ns", color)) %>% 
  dplyr::add_count(Compartment, sig_direction, gene) %>% 
  mutate(cons = ifelse(p.adjusted <= 0.05 & n == 3, "cons", "nc")) %>% 
  mutate(Compartment = fct_recode(factor(Compartment, levels = c("RS", "ES")),
                                  "Rhizosphere" = "RS",
                                  "Endosphere" = "ES")) %>% 
  ggplot(aes(baseMean, estimate, color = color, alpha = sig, shape = cons)) +
  geom_point(size = 1) +
  facet_grid(Compartment ~ Taxon, scales = "free_y") +
  scale_x_log10() +
  scale_shape_manual(values = c(16, 1)) +
  scale_alpha_manual(values = c(0.1, 0.7)) +
  scale_color_manual(values = c("#FF7F00", "grey20", "#984EA3", "#4DAF4A", "#377EB8")) +
  theme_minimal()

np_MA_plot
```

Why do this analysis? From our data so far, it appears that rice is an outlier. There must be OTUs that set it apart from the other plants. There are likely OTUs that are commonly enriched in rice compared to the other plants and OTUs that are commonly enriched in the other plants compared to rice. Let's go ahead and make some venn diagrams to figure this out.
```{r}
library(VennDiagram)
library(UpSetR)

deseq_results <- DESeq2_plant %>% 
  unnest(map(results, ~tidy(.))) %>% 
  separate(taxon, sep = "_", into = c("Taxon", "Compartment")) %>% 
  mutate(direction = ifelse(estimate < 0, "Weed", "Rice")) %>% 
  filter(p.adjusted <= 0.05) %>% 
  mutate(Compartment = gsub("ES", "Endosphere", Compartment),
         Compartment = gsub("RS", "Rhizosphere", Compartment)) %>% 
  dplyr::rename(variable = gene)

upset(deseq_results %>% 
        filter(Compartment == "Rhizosphere" & direction == "Rice") %>% 
        dplyr::count(variable, Taxon) %>% 
        spread(Taxon, n, fill = 0) %>% dplyr::select(-variable) %>% as.data.frame(),
      sets.bar.color = "black",
      order.by = "freq", 
      empty.intersections = "on")

upset(deseq_results %>% 
        filter(Compartment == "Endosphere" & direction == "Rice") %>% 
        dplyr::count(variable, Taxon) %>% 
        spread(Taxon, n, fill = 0) %>% dplyr::select(-variable) %>% as.data.frame(),
      sets.bar.color = "#56B4E9",
      order.by = "freq", 
      empty.intersections = "on")

upset(deseq_results %>% 
        filter(Compartment == "Rhizosphere" & direction == "Weed") %>% 
        dplyr::count(variable, Taxon) %>% 
        spread(Taxon, n, fill = 0) %>% dplyr::select(-variable) %>% as.data.frame(),
      sets.bar.color = "#56B4E9",
      order.by = "freq", 
      empty.intersections = "on")

upset(deseq_results %>% 
        filter(Compartment == "Endosphere" & direction == "Weed") %>% 
        dplyr::count(variable, Taxon) %>% 
        spread(Taxon, n, fill = 0) %>% dplyr::select(-variable) %>% as.data.frame(),
      sets.bar.color = "#56B4E9",
      order.by = "freq", 
      empty.intersections = "on")


```

There are quite a few OTUs specifically enriched in rice or the weeds, let's see what phyla these belong to.
```{r}
intersect_deseq <- DESeq2_plant %>% 
  unnest(map(results, ~tidy(.))) %>% 
  separate(taxon, sep = "_", into = c("Taxon", "Compartment")) %>% 
  mutate(direction = ifelse(estimate < 0, "Weed", "Rice")) %>% 
  filter(p.adjusted <= 0.05) %>% 
  dplyr::count(direction, Compartment, gene) %>% 
  dplyr::add_count(direction, Compartment) %>% 
  filter(n == 3) %>% 
  mutate(Compartment = gsub("ES", "Endosphere", Compartment),
         Compartment = gsub("RS", "Rhizosphere", Compartment)) %>% 
  dplyr::rename(variable = gene)

intersect_deseq %>% 
  inner_join(tax, by = "variable") %>% 
  dplyr::count(Compartment, direction, Phylum2) %>% 
  mutate(Compartment  = fct_relevel(factor(Compartment), "Rhizosphere", "Endosphere")) %>% 
  mutate(nnn2 = ifelse(direction == "Rice", nnn, -nnn)) %>% 
  group_by(Compartment) %>% 
  arrange(nnn2) %>% 
  mutate(order = 1:n()) %>% 
  mutate(direction_multiplier = ifelse(nnn2 > 0, 1, -1)) %>% 
  mutate(hjust = ifelse(nnn2 > 0, 0, 1)) %>% 
  mutate(location = ifelse(nnn >= abs(15), (1)*direction_multiplier, (nnn+0.5) * direction_multiplier)) %>% 
  mutate(color = ifelse(nnn >= abs(15), "white", "black")) %>% 
  ggplot(aes(order, nnn2, lable = "Phylum2", fill = factor(direction_multiplier))) +
  geom_bar(stat = "identity") +
  geom_text(aes(x = order, y = location, label = Phylum2, color = color, hjust = hjust), size = 3.5) +
  scale_color_manual(values = c("black", "white")) +
  facet_wrap( ~ Compartment, scales = "free") +
  theme_minimal() +
  theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), legend.position = "none") +
  coord_flip() +
  scale_fill_manual(values = c("black", "#4DAF4A"))
```
```{r}
nUniverse <- nExp %>% 
  inner_join(tax, by = "variable") %>% 
  group_by(variable) %>% 
  filter(sum(value > 0) / n() > 0.1) %>% 
  dplyr::count(variable, Phylum2) %>% ungroup() %>% 
  dplyr::count(Phylum2) %>% dplyr::rename(universe_phy = nn) %>% 
  mutate(universe_size = sum(universe_phy))
  
intersect_deseq %>% 
  inner_join(tax, by = "variable") %>% 
  dplyr::count(Compartment, direction, Phylum2) %>% 
  dplyr::rename(group_n = nnn) %>% 
  ungroup() %>% 
  group_by(Phylum2) %>%  
  group_by(direction, Compartment) %>% 
  mutate(group_size = sum(group_n)) %>% 
  inner_join(nUniverse, by = "Phylum2") %>% 
  mutate(phyp = phyper(group_n, universe_phy, universe_size - group_size, group_size, lower.tail = F)) %>% 
  mutate(p.adj = p.adjust(phyp, "fdr")) %>% 
  filter(p.adj <= 0.05) %>% 
  filter(group_n > 2)
```


One of the prevailing phyla coming out as being specifically enriched in rice is Euryarachaeota, which contains methanogenic archaea. Let's go a bit deeper here.
```{r}
rice_specific_mathanogen_otus <- intersect_deseq %>% 
  inner_join(tax, by = "variable") %>% 
  filter(grepl("Methano", Family) & direction == "Rice") %>% 
  dplyr::select(Compartment, variable) %>% 
  inner_join(tax, by = "variable")

methanogen_data <- nExp %>% 
  inner_join(tax, by = "variable") %>% 
  filter(grepl("Methano", Family)) %>% 
  group_by(variable, Compartment) %>%
  filter(sum(RA) > 0) %>% 
  summarise(mean_ab = mean(RA * 1000)) %>% 
  group_by(Compartment) %>% 
  arrange(-mean_ab) %>% 
  mutate(rank = 1:n())

mg_plot <- rbind(methanogen_data %>% filter(!variable%in%rice_specific_mathanogen_otus) %>%  mutate(color = "not_rice"),
                 methanogen_data %>% inner_join(rice_specific_mathanogen_otus, by = c("variable", "Compartment")) %>% mutate(color = "rice")) %>% 
  filter(Compartment != "Bulk Soil") %>% ungroup() %>% 
  mutate(Compartment = fct_relevel(factor(Compartment), "Rhizosphere", "Endosphere")) %>% 
  ggplot(aes(rank, mean_ab, color = color)) +
  geom_point(size = 2) +
  facet_grid(.~Compartment) +
  scale_y_log10() +
  scale_color_manual(values = c("black", "red")) +
  theme_minimal() +
  theme(text = element_text(size = 12))
```


What about methanotrophic eubacteria?
```{r}
methanotroph_otus <- intersect_deseq %>% 
  inner_join(tax, by = "variable") %>% 
  filter(grepl("Methyl", Family) & Genus != "Pleomorphomonas" & Family != "Methylophilaceae") 

methanotroph_data <- nExp %>% 
  inner_join(tax, by = "variable") %>% 
  filter(grepl("Methyl", Family) & Genus != "Pleomorphomonas" & Family != "Methylophilaceae") %>% 
  group_by(variable, Compartment) %>%
  filter(sum(value) > 0) %>% 
  summarise(mean_ab = mean(RA * 1000)) %>% 
  group_by(Compartment) %>% 
  arrange(-mean_ab) %>% 
  mutate(rank = 1:n())

mt_plot <- rbind(methanotroph_data %>% filter(!variable%in%unique(methanotroph_otus$variable)) %>%  mutate(color = "ns"),
                 methanotroph_data %>% inner_join(methanotroph_otus, by = c("variable", "Compartment")) %>% mutate(color = direction)) %>% 
  filter(Compartment != "Bulk Soil") %>% ungroup() %>% 
  mutate(Compartment = fct_relevel(factor(Compartment), "Rhizosphere", "Endosphere")) %>% 
  ggplot(aes(rank, mean_ab, color = color, shape = color)) +
  geom_point(size = 2) +
  facet_grid(.~Compartment) +
  scale_y_log10() +
  scale_shape_manual(values = c(16,1,16)) +
  scale_color_manual(values = c("black", "dodgerblue", "dodgerblue")) +
  theme_minimal() +
  theme(text = element_text(size = 12))
```

```{r}
DESeq2_comp <- nExp %>% 
   dplyr::select(SampleID, variable, value, Compartment, host_common_name) %>%
   filter(host_common_name %in% c("Rice", "Soil")) %>% 
   mutate(Compartment  = gsub("Bulk Soil", "BulkSoil", Compartment)) %>% 
   group_by(variable) %>% 
   filter(sum(value > 0) / n() > 0.1) %>% 
   mutate(host_common_name = relevel(factor(host_common_name), ref = "Rice")) %>% 
   mutate(group = factor(paste(Compartment))) %>% 
   group_by(group_var = "DESeq2") %>% 
   nest() %>% 
   mutate(DGEL = map(data, ~tidyDGEL(., value = "value", formula = "~group", method = "DESeq2"))) %>%
   mutate(dds = map(DGEL, ~DESeq(.))) %>% 
   mutate(Rhizosphere = map(dds, ~lfcShrink(., contrast = c("group", "Rhizosphere", "BulkSoil")))) %>% 
  mutate(Endosphere = map(dds, ~lfcShrink(., contrast = c("group", "Endosphere", "BulkSoil")))) %>% 
  dplyr::select(Rhizosphere, Endosphere) %>% 
   gather(key = Compartment, value = results) %>% 
  unnest(map(results, ~tidy(.)))
```

```{r}
DESeq2_comp %>% 
  dplyr::rename(variable = gene) %>% 
  inner_join(tax, by = "variable") %>% 
  filter(grepl("Methano", Family)) %>% 
  group_by(Compartment) %>% 
  mutate(p.adj = p.adjust(p.value, "BH")) %>% 
  filter(p.adj <= 0.05)
```

```{r}
rice_methanogen <- DESeq2_plant %>% 
  unnest(map(results, ~tidy(.))) %>% 
  separate(taxon, sep = "_", into = c("Taxon", "Compartment")) %>% 
  mutate(direction = ifelse(estimate < 0, "Weed", "Rice")) %>% 
  dplyr::rename(variable = gene) %>% 
  inner_join(tax) %>% 
  filter(grepl("Methan", Family)) 
  filter(direction == "Rice" & p.adjusted <= 0.05) %>%
  dplyr::count(Compartment, variable) %>% 
  select(-Compartment)
  
methano_RA <- nExp %>% 
  inner_join(tax, by = "variable") %>% 
  filter(grepl("Methano", Family)) %>% 
  group_by(variable) %>% 
  mutate(scaled = (RA - min(RA)) / (max(RA) - min(RA)))
  nExp_vsd %>%
  inner_join(tax) %>% 
  filter(grepl("Methano", Family)) %>% 
  group_by(variable, Compartment, host_common_name) %>% 
  filter(host_common_name != "Rice" | host_common_name == "Soil") 


DESeq2_comp %>% 
  dplyr::rename(variable = gene) %>% 
  inner_join(rice_methanogen, by = "variable") %>% 
  filter(p.adjusted <= 0.05)

nExp_vsd %>%
  inner_join(tax) %>% 
  filter(grepl("Methano", Family)) %>% 
  group_by(variable, Compartment, host_common_name) %>% 
  filter(host_common_name != "Rice" | host_common_name == "Soil") %>% 
  mutate(vsd = ifelse(vsd < 0, 0, vsd)) %>%
  group_by(1) %>% 
  nest() %>% 
  mutate(mod = map(data, ~aov(vsd ~ Compartment, .))) %>% 
  unnest(map(mod, ~tidy(TukeyHSD(.))))

r_meth <- nExp_vsd %>%
  inner_join(tax) %>% 
  filter(grepl("Methano", Family)) %>% 
  group_by(variable, Compartment, host_common_name) %>% 
  filter(host_common_name != "Rice" | host_common_name == "Soil") %>% 
  mutate(vsd = ifelse(vsd < 0, 0, vsd)) %>%
  #summarise(mean = mean(vsd)) %>% 
  ungroup() %>% 
  mutate(Compartment = fct_relevel(Compartment, "Bulk Soil", "Rhizosphere", "Endosphere")) %>% 
  ggplot(aes(Compartment, vsd)) +
  geom_boxplot(outlier.size = 0.1) +
  ylim(c(0,10)) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

methano_RA %>% 
  ungroup() %>% 
  mutate(Compartment = fct_relevel(Compartment, "Bulk Soil", "Rhizosphere", "Endosphere")) %>% 
  mutate(type = ifelse(host_common_name == "Rice", "Rice", "Other")) %>% 
  mutate(type = paste0(type, Compartment)) %>% 
  mutate(type = fct_relevel(type, "OtherBulk Soil", "OtherRhizosphere", "RiceRhizosphere", "OtherEndosphere", "RiceEndosphere")) %>% 
  ggplot(aes(type, scaled)) +
  geom_boxplot(outlier.size = 0.1) +
  #ylim(c(0,10)) +
  theme_minimal() +
  labs(x = "", y = "Scaled Abundance") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

methano_RA %>% 
  ungroup() %>% 
  mutate(Compartment = fct_relevel(Compartment, "Bulk Soil", "Rhizosphere", "Endosphere")) %>% 
  mutate(type = ifelse(host_common_name == "Rice", "Rice", "Other")) %>% 
  mutate(type = paste0(type, Compartment)) %>% 
  mutate(type = fct_relevel(type, "OtherBulk Soil", "OtherRhizosphere", "RiceRhizosphere", "OtherEndosphere", "RiceEndosphere")) %>% 
  group_by(1) %>% nest() %>% 
  mutate(mod = map(data, ~aov(scaled ~ type, .))) %>% 
  unnest(map(mod, ~tidy(TukeyHSD(.)))) %>% 
  select(comparison, estimate,adj.p.value)

nExp_vsd %>%
  inner_join(tax) %>% 
  filter(grepl("Methano", Family)) %>% 
  group_by(variable, Compartment, host_common_name, Order) %>% 
  #filter(host_common_name == "Rice" | host_common_name == "Soil") %>% 
  mutate(vsd = ifelse(vsd < 0, 0, vsd)) %>%
  #summarise(mean = mean(vsd)) %>% 
  group_by(Order) %>% 
  nest() %>% 
  mutate(mod = map(data, ~aov(vsd ~ Compartment, .))) %>% 
  unnest(map(mod, ~tidy(TukeyHSD(.))))

np_meth <- nExp_vsd %>%
  inner_join(tax) %>% 
  filter(grepl("Methano", Family)) %>% 
  group_by(variable, Compartment, host_common_name) %>% 
  filter(host_common_name == "Rice" | host_common_name == "Soil") %>% 
  mutate(vsd = ifelse(vsd < 0, 0, vsd)) %>%
  ungroup() %>% 
  mutate(Compartment = fct_relevel(Compartment, "Bulk Soil", "Rhizosphere", "Endosphere")) %>% 
  ggplot(aes(Compartment, vsd)) +
  geom_boxplot(outlier.size = 0.1) +
  ylim(c(0,10)) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

cowplot::plot_grid(r_meth, np_meth)
```


```{r}
library(gridExtra)
cowplot::plot_grid(mg_plot, mt_plot, nrow = 2)
```
```{r}
library(multcomp)

mg_mt_ratios <- rbind(nExp_vsd %>% inner_join(tax, by = "variable") %>% filter(grepl("Methano", Family)) %>% mutate(classification = "Methanogen"),
                      nExp_vsd %>% inner_join(tax, by = "variable") %>% filter(grepl("Methyl", Family) & Genus != "Pleomorphomonas" & Family != "Methylophilaceae")%>% mutate(classification = "Methanotroph")) %>% 
  group_by(SampleID, Compartment, host_common_name, classification) %>% 
  summarise(total = sum(vsd)) %>% 
  tidyr::spread(key = classification, value = total) %>% 
  mutate(ratio = Methanogen / Methanotroph)

mg_mt_ratios %>% 
  ungroup() %>% 
  filter(Compartment != "Bulk Soil") %>% 
  mutate(host_common_name = fct_relevel(host_common_name, "Rice")) %>% 
  mutate(Compartment = fct_relevel(factor(Compartment), "Rhizosphere", "Endosphere")) %>% 
  ggplot(aes(Compartment, log2(ratio), color = host_common_name)) +
  geom_hline(yintercept = 0, linetype = "dashed") +
  geom_point(position = position_jitterdodge(jitter.width = 0.1, jitter.height = 0), size = 2) +
  scale_color_manual(values = c("#4DAF4A", "#FF7F00", "#984EA3" , "#377EB8")) +
  theme_minimal() +
  theme(text = element_text(size = 12)) +
  labs(x = "", y = "log2(Methanogen / Methanotroph)")

mg_mt_ratios %>% 
  filter(Compartment != "Bulk Soil" & ratio > 0 ) %>% 
  group_by(Compartment) %>% 
  mutate(host_common_name = relevel(factor(host_common_name), ref = "Rice")) %>% 
  nest() %>% 
  mutate(model = map(data, ~lm(log2(ratio) ~ host_common_name, .))) %>% 
  unnest(map(model, ~tidy(.)))

mg_mt_ratios %>% 
  filter(Compartment != "Bulk Soil" & ratio > 0 ) %>% ungroup() %>% 
  mutate(host_common_name = relevel(factor(host_common_name), ref = "Rice")) %>% 
  group_by(1) %>% 
  nest() %>% 
  mutate(model = map(data, ~aov(log2(ratio) ~ Compartment:host_common_name, .))) %>% 
  unnest(map(model, ~tidy(TukeyHSD(.)))) %>% 
  filter(adj.p.value <= 0.05) %>% 
  filter(grepl("Rice", comparison))

```


## Comparisons to bulk soil
Looking at the above PCoA, the communities in the rice rhizosphere appear to be remarkably similar to bulk soil. Let's go more into depth and ask whether this is statistically true.
```{r}
bs_dists <- tidyMB::long_distance(nExp) %>% 
  filter(Compartment.x == "Bulk Soil" | Compartment.y == "Bulk Soil") %>% 
  rowwise() %>% 
  mutate(bs_column = min(c(Compartment.x, Compartment.y))) %>% 
  mutate(compartment_column = max(c(Compartment.x, Compartment.y))) %>% 
  mutate(taxon_name = min(c(host_common_name.x, host_common_name.y))) %>% 
  ungroup() %>% 
  mutate(taxon_name = fct_relevel(factor(taxon_name), "Rice", "Mudplantain", "Redstem", "Sedge", "Soil")) %>% 
  mutate(compartment_column = fct_relevel(factor(compartment_column), "Bulk Soil", "Rhizosphere", "Endosphere"))

bs_dists %>% 
  ggplot(aes(compartment_column, value, color = taxon_name)) +
  geom_point(position = position_jitterdodge(jitter.width = 0.3), alpha = 0.8) + 
  scale_color_manual(values = c("#4DAF4A", "#FF7F00", "#984EA3", "#377EB8", "#E41A1C")) +
  theme_minimal() +
  labs(x = "", y = "Bray Dissimilarity") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1), text = element_text(size = 12))
```
It looks like rice has a reduced dissimilarity to bulk soil compared to the other plant species. Let's check whether this is significant
```{r}
bs_dists %>% 
  filter(compartment_column == "Rhizosphere") %>% 
  mutate(taxon_name = relevel(taxon_name, ref = "Rice")) %>% 
  aov(value ~ taxon_name, .) %>% 
  TukeyHSD(.) %>% 
  tidy(.)
```
From this ANOVA result, it looks like rice rhizosphere microbiomes are significantly more similar to the bulk soil microbiome than the other plant species.





